<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Joel Jojo">

<title>Exercise 2</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="Exercise 2_files/libs/clipboard/clipboard.min.js"></script>
<script src="Exercise 2_files/libs/quarto-html/quarto.js"></script>
<script src="Exercise 2_files/libs/quarto-html/popper.min.js"></script>
<script src="Exercise 2_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="Exercise 2_files/libs/quarto-html/anchor.min.js"></script>
<link href="Exercise 2_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="Exercise 2_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="Exercise 2_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="Exercise 2_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="Exercise 2_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Exercise 2</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Joel Jojo </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="summary" class="level2">
<h2 class="anchored" data-anchor-id="summary">Summary</h2>
<p>Topics included in this exercise are data submission, package installation, data loading, cleaning, scaling, correlations, dimensionality reduction, sampling, and handling text datasets.</p>
<p>Submission: How to submit data for analysis.</p>
<p>Installing Required Packages: Installing and managing R packages.</p>
<p>Loading Data: Importing datasets into R.</p>
<p>Data Cleaning: Removing inconsistencies and errors in data.</p>
<p>Scaling and Normalization: Standardizing data for analysis.</p>
<p>Variable Correlations and Dimensionality Reduction: Analyzing feature relationships and reducing dimensions.</p>
<p>Sampling: Methods for selecting representative data subsets.</p>
<p>Handling Text Datasets: Techniques for working with text-based data.</p>
</section>
<section id="code" class="level2">
<h2 class="anchored" data-anchor-id="code">Code</h2>
<p>The below code installs the “pacman” package which helps in packet installation and management in R.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># First run this</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"pacman"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into 'C:/Users/Joel Jojo/AppData/Local/R/win-library/4.3'
(as 'lib' is unspecified)</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in contrib.url(repos, "source"): trying to use CRAN without setting a mirror</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pacman)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">p_load</span>(DBI, <span class="co"># DBI databases</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>       dlookr,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>       here, <span class="co"># Reproducible/ standard directories</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>       janitor,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>       RMySQL,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>       sampling,<span class="co"># Utilizing MySQL drivers</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>       tidymodels, <span class="co"># Tidyverse format modeling (e.g., lm())</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>       tidyverse, <span class="co"># Data wrangling, manipulation, visualization</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>       qqplotr,</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>       tm) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="loading-data" class="level4">
<h4 class="anchored" data-anchor-id="loading-data">LOADING DATA</h4>
<p><strong>In the below section we practice loading data from a .csv or .txt file making them available for analysis and other purposes.</strong></p>
<p>The below code reads the x.csv file into “data” data frame using the read_csv() function.</p>
<p>data |&gt; glimpse() line passes the data frame into the glimpse() function which then displays a concise summary of the data frame’s structure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"x.csv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 4 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (1): Name
dbl (2): ID, Age

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>data <span class="sc">|&gt;</span> <span class="fu">glimpse</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 4
Columns: 3
$ ID   &lt;dbl&gt; 12345, 12346, 12347, 12348
$ Name &lt;chr&gt; "Amy", "Ana", "June", "May"
$ Age  &lt;dbl&gt; 13, 14, 25, 21</code></pre>
</div>
</div>
<p>The below code reads a tab delimited file from the “data” directory using read_delim() function. Then the glimpse function produces a summary of the data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"x.tsv"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 4 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: "\t"
chr (1): Name
dbl (2): ID, Age

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>data <span class="sc">|&gt;</span> <span class="fu">glimpse</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 4
Columns: 3
$ ID   &lt;dbl&gt; 12345, 12346, 12347, 12348
$ Name &lt;chr&gt; "Amy", "Ana", "June", "May"
$ Age  &lt;dbl&gt; 13, 14, NA, 21</code></pre>
</div>
</div>
<p>Unable to recreate <strong>Importing data from MySQL database</strong> section since the connection to MySQL isn’t getting established.<br>
<br>
</p>
</section>
<section id="data-cleaning" class="level4">
<h4 class="anchored" data-anchor-id="data-cleaning">DATA CLEANING</h4>
<p><strong>In the below section we will be experimenting with data cleaning methods wherein we identify and correct errors, inconsistencies and missing values.</strong></p>
<p><br>
The below code loads the wide.txt file into the “wide” data frame and using the delimiter ” ” it skips the first row of data. The col_names parameter provides custom column names to the data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>wide <span class="ot">&lt;-</span> <span class="fu">read_delim</span>(<span class="fu">here</span>(<span class="st">"data"</span>, <span class="st">"wide.txt"</span>), <span class="at">delim =</span> <span class="st">" "</span>, <span class="at">skip =</span> <span class="dv">1</span>, <span class="at">col_names =</span> <span class="fu">c</span>(<span class="st">"Name"</span>, <span class="st">"Math"</span>, <span class="st">"English"</span>, <span class="st">"Degree_Year"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 3 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: " "
chr (2): Name, Degree_Year
dbl (2): Math, English

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>The below code converts the data frame from a wide format to a long format.</p>
<p>cols = c(Math, English) line helps pivot the columns (Math, English) from a wide to a long format.</p>
<p>names_to = “Subject” line specifies that there will be a new column by the name “Subject” which will have the above columns as values.</p>
<p>values_to = “Grade” line specifies that there will be a new column by the name “Grade” which will have the grades corresponding to each subject.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>long <span class="ot">&lt;-</span> wide <span class="sc">|&gt;</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(Math, English),</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">names_to =</span> <span class="st">"Subject"</span>, </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">values_to =</span> <span class="st">"Grade"</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>long</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 4
  Name  Degree_Year Subject Grade
  &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;   &lt;dbl&gt;
1 Anna  Bio_2014    Math       86
2 Anna  Bio_2014    English    90
3 John  Math_2013   Math       43
4 John  Math_2013   English    75
5 Cath  Bio_2012    Math       80
6 Cath  Bio_2012    English    82</code></pre>
</div>
</div>
<p>In the below code we change the data frame from a long format to a wide format.</p>
<p>names_from = Subject argument specifies that the values in “Subject” column can be used as columns in the new data frame.</p>
<p>values_from = Grade argument specifies that the values in the “Grade” column should be used as the values for the new columns in the data frame.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>wide <span class="ot">&lt;-</span> long <span class="sc">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> Subject, <span class="at">values_from =</span> Grade)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>wide</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  Name  Degree_Year  Math English
  &lt;chr&gt; &lt;chr&gt;       &lt;dbl&gt;   &lt;dbl&gt;
1 Anna  Bio_2014       86      90
2 John  Math_2013      43      75
3 Cath  Bio_2012       80      82</code></pre>
</div>
</div>
<p>The below code uses the separate() function to separate the column “Degree_Year” into two new columns “Degree” and “Year”.</p>
<p>The values in the column will be separated using the underscore (_) as the separator.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>clean <span class="ot">&lt;-</span> long <span class="sc">%&gt;%</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(Degree_Year, <span class="fu">c</span>(<span class="st">"Degree"</span>, <span class="st">"Year"</span>), <span class="at">sep =</span> <span class="st">"_"</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>clean</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  Name  Degree Year  Subject Grade
  &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt;
1 Anna  Bio    2014  Math       86
2 Anna  Bio    2014  English    90
3 John  Math   2013  Math       43
4 John  Math   2013  English    75
5 Cath  Bio    2012  Math       80
6 Cath  Bio    2012  English    82</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"lubridate"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'lubridate' is in use and will not be installed</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lubridate)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The below code enters date in different formats into mixed.dates data frame.</p>
<p>It is then put into the ymd() function which then converts all the dates to a year-month-date format</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>mixed.dates <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">20140123</span>, <span class="st">"2019-12-12"</span>, <span class="st">"2009/5/1"</span>,</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a> <span class="st">"measured on 2002-12-06"</span>, <span class="st">"2018-7/16"</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>clean.dates <span class="ot">&lt;-</span> <span class="fu">ymd</span>(mixed.dates) <span class="co">#convert to year-month-day format</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>clean.dates</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2014-01-23" "2019-12-12" "2009-05-01" "2002-12-06" "2018-07-16"</code></pre>
</div>
</div>
<p>The below code uses various functions of the lubridate package to clean and extract date information.</p>
<p>The wday() function extracts the numeric representation of the day of the week(1 for sunday, 2 for monday etc.). If the same function has the label argument set to TRUE then it provides the label for the day of the week.<br>
year() function extracts the year information from the date.</p>
<p>month() function extracts the month information from the date.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data.frame</span>(<span class="at">Dates =</span> clean.dates, <span class="at">WeekDay =</span> <span class="fu">wday</span>(clean.dates), <span class="at">nWeekDay =</span> <span class="fu">wday</span>(clean.dates, <span class="at">label =</span> <span class="cn">TRUE</span>), <span class="at">Year =</span> <span class="fu">year</span>(clean.dates), <span class="at">Month =</span> <span class="fu">month</span>(clean.dates, <span class="at">label =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Dates WeekDay nWeekDay Year Month
1 2014-01-23       5      Thu 2014   Jan
2 2019-12-12       5      Thu 2019   Dec
3 2009-05-01       6      Fri 2009   May
4 2002-12-06       6      Fri 2002   Dec
5 2018-07-16       2      Mon 2018   Jul</code></pre>
</div>
</div>
<p>The below code contains assigns the Asia/Shanghai equivalent value of “20190203 03:00:03” to date.time</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>date.time <span class="ot">&lt;-</span> <span class="fu">ymd_hms</span>(<span class="st">"20190203 03:00:03"</span>, <span class="at">tz=</span><span class="st">"Asia/Shanghai"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The below code changes the date.time value using the with_tz() function to America/Phoenix time.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">with_tz</span>(date.time, <span class="at">tz=</span><span class="st">"America/Phoenix"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2019-02-02 12:00:03 MST"</code></pre>
</div>
</div>
<p>The below code utilizes the force_tz() function to forcefully change the date.time value from Phoenix time to Turkey time</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">force_tz</span>(date.time, <span class="st">"Turkey"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "2019-02-03 03:00:03 +03"</code></pre>
</div>
</div>
<p>The below code returns a list of timezones which can be used further while working on timezones.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">OlsonNames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "Africa/Abidjan"                   "Africa/Accra"                    
  [3] "Africa/Addis_Ababa"               "Africa/Algiers"                  
  [5] "Africa/Asmara"                    "Africa/Asmera"                   
  [7] "Africa/Bamako"                    "Africa/Bangui"                   
  [9] "Africa/Banjul"                    "Africa/Bissau"                   
 [11] "Africa/Blantyre"                  "Africa/Brazzaville"              
 [13] "Africa/Bujumbura"                 "Africa/Cairo"                    
 [15] "Africa/Casablanca"                "Africa/Ceuta"                    
 [17] "Africa/Conakry"                   "Africa/Dakar"                    
 [19] "Africa/Dar_es_Salaam"             "Africa/Djibouti"                 
 [21] "Africa/Douala"                    "Africa/El_Aaiun"                 
 [23] "Africa/Freetown"                  "Africa/Gaborone"                 
 [25] "Africa/Harare"                    "Africa/Johannesburg"             
 [27] "Africa/Juba"                      "Africa/Kampala"                  
 [29] "Africa/Khartoum"                  "Africa/Kigali"                   
 [31] "Africa/Kinshasa"                  "Africa/Lagos"                    
 [33] "Africa/Libreville"                "Africa/Lome"                     
 [35] "Africa/Luanda"                    "Africa/Lubumbashi"               
 [37] "Africa/Lusaka"                    "Africa/Malabo"                   
 [39] "Africa/Maputo"                    "Africa/Maseru"                   
 [41] "Africa/Mbabane"                   "Africa/Mogadishu"                
 [43] "Africa/Monrovia"                  "Africa/Nairobi"                  
 [45] "Africa/Ndjamena"                  "Africa/Niamey"                   
 [47] "Africa/Nouakchott"                "Africa/Ouagadougou"              
 [49] "Africa/Porto-Novo"                "Africa/Sao_Tome"                 
 [51] "Africa/Timbuktu"                  "Africa/Tripoli"                  
 [53] "Africa/Tunis"                     "Africa/Windhoek"                 
 [55] "America/Adak"                     "America/Anchorage"               
 [57] "America/Anguilla"                 "America/Antigua"                 
 [59] "America/Araguaina"                "America/Argentina/Buenos_Aires"  
 [61] "America/Argentina/Catamarca"      "America/Argentina/ComodRivadavia"
 [63] "America/Argentina/Cordoba"        "America/Argentina/Jujuy"         
 [65] "America/Argentina/La_Rioja"       "America/Argentina/Mendoza"       
 [67] "America/Argentina/Rio_Gallegos"   "America/Argentina/Salta"         
 [69] "America/Argentina/San_Juan"       "America/Argentina/San_Luis"      
 [71] "America/Argentina/Tucuman"        "America/Argentina/Ushuaia"       
 [73] "America/Aruba"                    "America/Asuncion"                
 [75] "America/Atikokan"                 "America/Atka"                    
 [77] "America/Bahia"                    "America/Bahia_Banderas"          
 [79] "America/Barbados"                 "America/Belem"                   
 [81] "America/Belize"                   "America/Blanc-Sablon"            
 [83] "America/Boa_Vista"                "America/Bogota"                  
 [85] "America/Boise"                    "America/Buenos_Aires"            
 [87] "America/Cambridge_Bay"            "America/Campo_Grande"            
 [89] "America/Cancun"                   "America/Caracas"                 
 [91] "America/Catamarca"                "America/Cayenne"                 
 [93] "America/Cayman"                   "America/Chicago"                 
 [95] "America/Chihuahua"                "America/Ciudad_Juarez"           
 [97] "America/Coral_Harbour"            "America/Cordoba"                 
 [99] "America/Costa_Rica"               "America/Creston"                 
[101] "America/Cuiaba"                   "America/Curacao"                 
[103] "America/Danmarkshavn"             "America/Dawson"                  
[105] "America/Dawson_Creek"             "America/Denver"                  
[107] "America/Detroit"                  "America/Dominica"                
[109] "America/Edmonton"                 "America/Eirunepe"                
[111] "America/El_Salvador"              "America/Ensenada"                
[113] "America/Fort_Nelson"              "America/Fort_Wayne"              
[115] "America/Fortaleza"                "America/Glace_Bay"               
[117] "America/Godthab"                  "America/Goose_Bay"               
[119] "America/Grand_Turk"               "America/Grenada"                 
[121] "America/Guadeloupe"               "America/Guatemala"               
[123] "America/Guayaquil"                "America/Guyana"                  
[125] "America/Halifax"                  "America/Havana"                  
[127] "America/Hermosillo"               "America/Indiana/Indianapolis"    
[129] "America/Indiana/Knox"             "America/Indiana/Marengo"         
[131] "America/Indiana/Petersburg"       "America/Indiana/Tell_City"       
[133] "America/Indiana/Vevay"            "America/Indiana/Vincennes"       
[135] "America/Indiana/Winamac"          "America/Indianapolis"            
[137] "America/Inuvik"                   "America/Iqaluit"                 
[139] "America/Jamaica"                  "America/Jujuy"                   
[141] "America/Juneau"                   "America/Kentucky/Louisville"     
[143] "America/Kentucky/Monticello"      "America/Knox_IN"                 
[145] "America/Kralendijk"               "America/La_Paz"                  
[147] "America/Lima"                     "America/Los_Angeles"             
[149] "America/Louisville"               "America/Lower_Princes"           
[151] "America/Maceio"                   "America/Managua"                 
[153] "America/Manaus"                   "America/Marigot"                 
[155] "America/Martinique"               "America/Matamoros"               
[157] "America/Mazatlan"                 "America/Mendoza"                 
[159] "America/Menominee"                "America/Merida"                  
[161] "America/Metlakatla"               "America/Mexico_City"             
[163] "America/Miquelon"                 "America/Moncton"                 
[165] "America/Monterrey"                "America/Montevideo"              
[167] "America/Montreal"                 "America/Montserrat"              
[169] "America/Nassau"                   "America/New_York"                
[171] "America/Nipigon"                  "America/Nome"                    
[173] "America/Noronha"                  "America/North_Dakota/Beulah"     
[175] "America/North_Dakota/Center"      "America/North_Dakota/New_Salem"  
[177] "America/Nuuk"                     "America/Ojinaga"                 
[179] "America/Panama"                   "America/Pangnirtung"             
[181] "America/Paramaribo"               "America/Phoenix"                 
[183] "America/Port-au-Prince"           "America/Port_of_Spain"           
[185] "America/Porto_Acre"               "America/Porto_Velho"             
[187] "America/Puerto_Rico"              "America/Punta_Arenas"            
[189] "America/Rainy_River"              "America/Rankin_Inlet"            
[191] "America/Recife"                   "America/Regina"                  
[193] "America/Resolute"                 "America/Rio_Branco"              
[195] "America/Rosario"                  "America/Santa_Isabel"            
[197] "America/Santarem"                 "America/Santiago"                
[199] "America/Santo_Domingo"            "America/Sao_Paulo"               
[201] "America/Scoresbysund"             "America/Shiprock"                
[203] "America/Sitka"                    "America/St_Barthelemy"           
[205] "America/St_Johns"                 "America/St_Kitts"                
[207] "America/St_Lucia"                 "America/St_Thomas"               
[209] "America/St_Vincent"               "America/Swift_Current"           
[211] "America/Tegucigalpa"              "America/Thule"                   
[213] "America/Thunder_Bay"              "America/Tijuana"                 
[215] "America/Toronto"                  "America/Tortola"                 
[217] "America/Vancouver"                "America/Virgin"                  
[219] "America/Whitehorse"               "America/Winnipeg"                
[221] "America/Yakutat"                  "America/Yellowknife"             
[223] "Antarctica/Casey"                 "Antarctica/Davis"                
[225] "Antarctica/DumontDUrville"        "Antarctica/Macquarie"            
[227] "Antarctica/Mawson"                "Antarctica/McMurdo"              
[229] "Antarctica/Palmer"                "Antarctica/Rothera"              
[231] "Antarctica/South_Pole"            "Antarctica/Syowa"                
[233] "Antarctica/Troll"                 "Antarctica/Vostok"               
[235] "Arctic/Longyearbyen"              "Asia/Aden"                       
[237] "Asia/Almaty"                      "Asia/Amman"                      
[239] "Asia/Anadyr"                      "Asia/Aqtau"                      
[241] "Asia/Aqtobe"                      "Asia/Ashgabat"                   
[243] "Asia/Ashkhabad"                   "Asia/Atyrau"                     
[245] "Asia/Baghdad"                     "Asia/Bahrain"                    
[247] "Asia/Baku"                        "Asia/Bangkok"                    
[249] "Asia/Barnaul"                     "Asia/Beirut"                     
[251] "Asia/Bishkek"                     "Asia/Brunei"                     
[253] "Asia/Calcutta"                    "Asia/Chita"                      
[255] "Asia/Choibalsan"                  "Asia/Chongqing"                  
[257] "Asia/Chungking"                   "Asia/Colombo"                    
[259] "Asia/Dacca"                       "Asia/Damascus"                   
[261] "Asia/Dhaka"                       "Asia/Dili"                       
[263] "Asia/Dubai"                       "Asia/Dushanbe"                   
[265] "Asia/Famagusta"                   "Asia/Gaza"                       
[267] "Asia/Harbin"                      "Asia/Hebron"                     
[269] "Asia/Ho_Chi_Minh"                 "Asia/Hong_Kong"                  
[271] "Asia/Hovd"                        "Asia/Irkutsk"                    
[273] "Asia/Istanbul"                    "Asia/Jakarta"                    
[275] "Asia/Jayapura"                    "Asia/Jerusalem"                  
[277] "Asia/Kabul"                       "Asia/Kamchatka"                  
[279] "Asia/Karachi"                     "Asia/Kashgar"                    
[281] "Asia/Kathmandu"                   "Asia/Katmandu"                   
[283] "Asia/Khandyga"                    "Asia/Kolkata"                    
[285] "Asia/Krasnoyarsk"                 "Asia/Kuala_Lumpur"               
[287] "Asia/Kuching"                     "Asia/Kuwait"                     
[289] "Asia/Macao"                       "Asia/Macau"                      
[291] "Asia/Magadan"                     "Asia/Makassar"                   
[293] "Asia/Manila"                      "Asia/Muscat"                     
[295] "Asia/Nicosia"                     "Asia/Novokuznetsk"               
[297] "Asia/Novosibirsk"                 "Asia/Omsk"                       
[299] "Asia/Oral"                        "Asia/Phnom_Penh"                 
[301] "Asia/Pontianak"                   "Asia/Pyongyang"                  
[303] "Asia/Qatar"                       "Asia/Qostanay"                   
[305] "Asia/Qyzylorda"                   "Asia/Rangoon"                    
[307] "Asia/Riyadh"                      "Asia/Saigon"                     
[309] "Asia/Sakhalin"                    "Asia/Samarkand"                  
[311] "Asia/Seoul"                       "Asia/Shanghai"                   
[313] "Asia/Singapore"                   "Asia/Srednekolymsk"              
[315] "Asia/Taipei"                      "Asia/Tashkent"                   
[317] "Asia/Tbilisi"                     "Asia/Tehran"                     
[319] "Asia/Tel_Aviv"                    "Asia/Thimbu"                     
[321] "Asia/Thimphu"                     "Asia/Tokyo"                      
[323] "Asia/Tomsk"                       "Asia/Ujung_Pandang"              
[325] "Asia/Ulaanbaatar"                 "Asia/Ulan_Bator"                 
[327] "Asia/Urumqi"                      "Asia/Ust-Nera"                   
[329] "Asia/Vientiane"                   "Asia/Vladivostok"                
[331] "Asia/Yakutsk"                     "Asia/Yangon"                     
[333] "Asia/Yekaterinburg"               "Asia/Yerevan"                    
[335] "Atlantic/Azores"                  "Atlantic/Bermuda"                
[337] "Atlantic/Canary"                  "Atlantic/Cape_Verde"             
[339] "Atlantic/Faeroe"                  "Atlantic/Faroe"                  
[341] "Atlantic/Jan_Mayen"               "Atlantic/Madeira"                
[343] "Atlantic/Reykjavik"               "Atlantic/South_Georgia"          
[345] "Atlantic/St_Helena"               "Atlantic/Stanley"                
[347] "Australia/ACT"                    "Australia/Adelaide"              
[349] "Australia/Brisbane"               "Australia/Broken_Hill"           
[351] "Australia/Canberra"               "Australia/Currie"                
[353] "Australia/Darwin"                 "Australia/Eucla"                 
[355] "Australia/Hobart"                 "Australia/LHI"                   
[357] "Australia/Lindeman"               "Australia/Lord_Howe"             
[359] "Australia/Melbourne"              "Australia/North"                 
[361] "Australia/NSW"                    "Australia/Perth"                 
[363] "Australia/Queensland"             "Australia/South"                 
[365] "Australia/Sydney"                 "Australia/Tasmania"              
[367] "Australia/Victoria"               "Australia/West"                  
[369] "Australia/Yancowinna"             "Brazil/Acre"                     
[371] "Brazil/DeNoronha"                 "Brazil/East"                     
[373] "Brazil/West"                      "Canada/Atlantic"                 
[375] "Canada/Central"                   "Canada/Eastern"                  
[377] "Canada/Mountain"                  "Canada/Newfoundland"             
[379] "Canada/Pacific"                   "Canada/Saskatchewan"             
[381] "Canada/Yukon"                     "CET"                             
[383] "Chile/Continental"                "Chile/EasterIsland"              
[385] "CST6CDT"                          "Cuba"                            
[387] "EET"                              "Egypt"                           
[389] "Eire"                             "EST"                             
[391] "EST5EDT"                          "Etc/GMT"                         
[393] "Etc/GMT-0"                        "Etc/GMT-1"                       
[395] "Etc/GMT-10"                       "Etc/GMT-11"                      
[397] "Etc/GMT-12"                       "Etc/GMT-13"                      
[399] "Etc/GMT-14"                       "Etc/GMT-2"                       
[401] "Etc/GMT-3"                        "Etc/GMT-4"                       
[403] "Etc/GMT-5"                        "Etc/GMT-6"                       
[405] "Etc/GMT-7"                        "Etc/GMT-8"                       
[407] "Etc/GMT-9"                        "Etc/GMT+0"                       
[409] "Etc/GMT+1"                        "Etc/GMT+10"                      
[411] "Etc/GMT+11"                       "Etc/GMT+12"                      
[413] "Etc/GMT+2"                        "Etc/GMT+3"                       
[415] "Etc/GMT+4"                        "Etc/GMT+5"                       
[417] "Etc/GMT+6"                        "Etc/GMT+7"                       
[419] "Etc/GMT+8"                        "Etc/GMT+9"                       
[421] "Etc/GMT0"                         "Etc/Greenwich"                   
[423] "Etc/UCT"                          "Etc/Universal"                   
[425] "Etc/UTC"                          "Etc/Zulu"                        
[427] "Europe/Amsterdam"                 "Europe/Andorra"                  
[429] "Europe/Astrakhan"                 "Europe/Athens"                   
[431] "Europe/Belfast"                   "Europe/Belgrade"                 
[433] "Europe/Berlin"                    "Europe/Bratislava"               
[435] "Europe/Brussels"                  "Europe/Bucharest"                
[437] "Europe/Budapest"                  "Europe/Busingen"                 
[439] "Europe/Chisinau"                  "Europe/Copenhagen"               
[441] "Europe/Dublin"                    "Europe/Gibraltar"                
[443] "Europe/Guernsey"                  "Europe/Helsinki"                 
[445] "Europe/Isle_of_Man"               "Europe/Istanbul"                 
[447] "Europe/Jersey"                    "Europe/Kaliningrad"              
[449] "Europe/Kiev"                      "Europe/Kirov"                    
[451] "Europe/Kyiv"                      "Europe/Lisbon"                   
[453] "Europe/Ljubljana"                 "Europe/London"                   
[455] "Europe/Luxembourg"                "Europe/Madrid"                   
[457] "Europe/Malta"                     "Europe/Mariehamn"                
[459] "Europe/Minsk"                     "Europe/Monaco"                   
[461] "Europe/Moscow"                    "Europe/Nicosia"                  
[463] "Europe/Oslo"                      "Europe/Paris"                    
[465] "Europe/Podgorica"                 "Europe/Prague"                   
[467] "Europe/Riga"                      "Europe/Rome"                     
[469] "Europe/Samara"                    "Europe/San_Marino"               
[471] "Europe/Sarajevo"                  "Europe/Saratov"                  
[473] "Europe/Simferopol"                "Europe/Skopje"                   
[475] "Europe/Sofia"                     "Europe/Stockholm"                
[477] "Europe/Tallinn"                   "Europe/Tirane"                   
[479] "Europe/Tiraspol"                  "Europe/Ulyanovsk"                
[481] "Europe/Uzhgorod"                  "Europe/Vaduz"                    
[483] "Europe/Vatican"                   "Europe/Vienna"                   
[485] "Europe/Vilnius"                   "Europe/Volgograd"                
[487] "Europe/Warsaw"                    "Europe/Zagreb"                   
[489] "Europe/Zaporozhye"                "Europe/Zurich"                   
[491] "GB"                               "GB-Eire"                         
[493] "GMT"                              "GMT-0"                           
[495] "GMT+0"                            "GMT0"                            
[497] "Greenwich"                        "Hongkong"                        
[499] "HST"                              "Iceland"                         
[501] "Indian/Antananarivo"              "Indian/Chagos"                   
[503] "Indian/Christmas"                 "Indian/Cocos"                    
[505] "Indian/Comoro"                    "Indian/Kerguelen"                
[507] "Indian/Mahe"                      "Indian/Maldives"                 
[509] "Indian/Mauritius"                 "Indian/Mayotte"                  
[511] "Indian/Reunion"                   "Iran"                            
[513] "Israel"                           "Jamaica"                         
[515] "Japan"                            "Kwajalein"                       
[517] "Libya"                            "MET"                             
[519] "Mexico/BajaNorte"                 "Mexico/BajaSur"                  
[521] "Mexico/General"                   "MST"                             
[523] "MST7MDT"                          "Navajo"                          
[525] "NZ"                               "NZ-CHAT"                         
[527] "Pacific/Apia"                     "Pacific/Auckland"                
[529] "Pacific/Bougainville"             "Pacific/Chatham"                 
[531] "Pacific/Chuuk"                    "Pacific/Easter"                  
[533] "Pacific/Efate"                    "Pacific/Enderbury"               
[535] "Pacific/Fakaofo"                  "Pacific/Fiji"                    
[537] "Pacific/Funafuti"                 "Pacific/Galapagos"               
[539] "Pacific/Gambier"                  "Pacific/Guadalcanal"             
[541] "Pacific/Guam"                     "Pacific/Honolulu"                
[543] "Pacific/Johnston"                 "Pacific/Kanton"                  
[545] "Pacific/Kiritimati"               "Pacific/Kosrae"                  
[547] "Pacific/Kwajalein"                "Pacific/Majuro"                  
[549] "Pacific/Marquesas"                "Pacific/Midway"                  
[551] "Pacific/Nauru"                    "Pacific/Niue"                    
[553] "Pacific/Norfolk"                  "Pacific/Noumea"                  
[555] "Pacific/Pago_Pago"                "Pacific/Palau"                   
[557] "Pacific/Pitcairn"                 "Pacific/Pohnpei"                 
[559] "Pacific/Ponape"                   "Pacific/Port_Moresby"            
[561] "Pacific/Rarotonga"                "Pacific/Saipan"                  
[563] "Pacific/Samoa"                    "Pacific/Tahiti"                  
[565] "Pacific/Tarawa"                   "Pacific/Tongatapu"               
[567] "Pacific/Truk"                     "Pacific/Wake"                    
[569] "Pacific/Wallis"                   "Pacific/Yap"                     
[571] "Poland"                           "Portugal"                        
[573] "PRC"                              "PST8PDT"                         
[575] "ROC"                              "ROK"                             
[577] "Singapore"                        "Turkey"                          
[579] "UCT"                              "Universal"                       
[581] "US/Alaska"                        "US/Aleutian"                     
[583] "US/Arizona"                       "US/Central"                      
[585] "US/East-Indiana"                  "US/Eastern"                      
[587] "US/Hawaii"                        "US/Indiana-Starke"               
[589] "US/Michigan"                      "US/Mountain"                     
[591] "US/Pacific"                       "US/Samoa"                        
[593] "UTC"                              "W-SU"                            
[595] "WET"                              "Zulu"                            
attr(,"Version")
[1] "2023c"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(stringr)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The variable uci.repo stores a character string representing the base URL of the UCI Machine Learning Repository. The base URL hosts different ML datasets.</p>
<p>The “dataset” variable stores a dataset, specifically “audiology” from the ML repository.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>uci.repo <span class="ot">&lt;-</span><span class="st">"http://archive.ics.uci.edu/ml/machine-learning-databases/"</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>dataset <span class="ot">&lt;-</span> <span class="st">"audiology/audiology.standardized"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The below code uses the str_c() function to concatenate three strings together. This helps achieve complete access to the data and associated variable names for the specified dataset in the UCI Machine Learning directory.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>dataF <span class="ot">&lt;-</span> <span class="fu">str_c</span>(uci.repo, dataset, <span class="st">".data"</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>namesF <span class="ot">&lt;-</span> <span class="fu">str_c</span>(uci.repo, dataset, <span class="st">".names"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>dataF</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "http://archive.ics.uci.edu/ml/machine-learning-databases/audiology/audiology.standardized.data"</code></pre>
</div>
</div>
<p>This line of code reads the first row of data without considering them as columns (col_names= FALSE) and the missing values “?” are replaced with NA. The url() function helps create a connection to the specified URL.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="fu">url</span>(dataF), <span class="at">col_names =</span> <span class="cn">FALSE</span>, <span class="at">na=</span><span class="st">"?"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 200 Columns: 71
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (11): X2, X4, X5, X6, X8, X59, X60, X64, X66, X70, X71
lgl (60): X1, X3, X7, X9, X10, X11, X12, X13, X14, X15, X16, X17, X18, X19, ...

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
</div>
<p>The below code returns the dimensions(number of rows, number of columns) of the “data” data frame using the dim() function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 200  71</code></pre>
</div>
</div>
<p>The below code reads the text content from the URL provided as an argument and stores it in the “lines” variable of which a few lines are displayed using the head() function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>lines <span class="ot">&lt;-</span> <span class="fu">read_lines</span>(<span class="fu">url</span>(namesF))</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>lines <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "WARNING: This database should be credited to the original owner whenever"
[2] "         used for any publication whatsoever."                           
[3] ""                                                                        
[4] "1. Title: Standardized Audiology Database "                              
[5] ""                                                                        
[6] "2. Sources:"                                                             </code></pre>
</div>
</div>
<p>The below code reads the lines 67 through 135 and displays it through the “names” variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> lines[<span class="dv">67</span><span class="sc">:</span><span class="dv">135</span>]</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "   age_gt_60:\t\t     f, t."                                          
 [2] "   air():\t\t     mild,moderate,severe,normal,profound."              
 [3] "   airBoneGap:\t\t     f, t."                                         
 [4] "   ar_c():\t\t     normal,elevated,absent."                           
 [5] "   ar_u():\t\t     normal,absent,elevated."                           
 [6] "   bone():\t\t     mild,moderate,normal,unmeasured."                  
 [7] "   boneAbnormal:\t     f, t."                                         
 [8] "   bser():\t\t     normal,degraded."                                  
 [9] "   history_buzzing:\t     f, t."                                      
[10] "   history_dizziness:\t     f, t."                                    
[11] "   history_fluctuating:\t     f, t."                                  
[12] "   history_fullness:\t     f, t."                                     
[13] "   history_heredity:\t     f, t."                                     
[14] "   history_nausea:\t     f, t."                                       
[15] "   history_noise:\t     f, t."                                        
[16] "   history_recruitment:\t     f, t."                                  
[17] "   history_ringing:\t     f, t."                                      
[18] "   history_roaring:\t     f, t."                                      
[19] "   history_vomiting:\t     f, t."                                     
[20] "   late_wave_poor:\t     f, t."                                       
[21] "   m_at_2k:\t\t     f, t."                                            
[22] "   m_cond_lt_1k:\t     f, t."                                         
[23] "   m_gt_1k:\t\t     f, t."                                            
[24] "   m_m_gt_2k:\t\t     f, t."                                          
[25] "   m_m_sn:\t\t     f, t."                                             
[26] "   m_m_sn_gt_1k:\t     f, t."                                         
[27] "   m_m_sn_gt_2k:\t     f, t."                                         
[28] "   m_m_sn_gt_500:\t     f, t."                                        
[29] "   m_p_sn_gt_2k:\t     f, t."                                         
[30] "   m_s_gt_500:\t\t     f, t."                                         
[31] "   m_s_sn:\t\t     f, t."                                             
[32] "   m_s_sn_gt_1k:\t     f, t."                                         
[33] "   m_s_sn_gt_2k:\t     f, t."                                         
[34] "   m_s_sn_gt_3k:\t     f, t."                                         
[35] "   m_s_sn_gt_4k:\t     f, t."                                         
[36] "   m_sn_2_3k:\t\t     f, t."                                          
[37] "   m_sn_gt_1k:\t\t     f, t."                                         
[38] "   m_sn_gt_2k:\t\t     f, t."                                         
[39] "   m_sn_gt_3k:\t\t     f, t."                                         
[40] "   m_sn_gt_4k:\t\t     f, t."                                         
[41] "   m_sn_gt_500:\t\t     f, t."                                        
[42] "   m_sn_gt_6k:\t\t     f, t."                                         
[43] "   m_sn_lt_1k:\t\t     f, t. "                                        
[44] "   m_sn_lt_2k:\t\t     f, t."                                         
[45] "   m_sn_lt_3k:\t\t     f, t."                                         
[46] "   middle_wave_poor:\t     f, t."                                     
[47] "   mod_gt_4k:\t\t     f, t."                                          
[48] "   mod_mixed:\t\t     f, t."                                          
[49] "   mod_s_mixed:\t\t     f, t."                                        
[50] "   mod_s_sn_gt_500:\t     f, t."                                      
[51] "   mod_sn:\t\t     f, t."                                             
[52] "   mod_sn_gt_1k:\t     f, t."                                         
[53] "   mod_sn_gt_2k:\t     f, t."                                         
[54] "   mod_sn_gt_3k:\t     f, t."                                         
[55] "   mod_sn_gt_4k:\t     f, t."                                         
[56] "   mod_sn_gt_500:\t     f, t."                                        
[57] "   notch_4k:\t\t     f, t."                                           
[58] "   notch_at_4k:\t\t     f, t."                                        
[59] "   o_ar_c():\t\t     normal,elevated,absent."                         
[60] "   o_ar_u():\t\t     normal,absent,elevated."                         
[61] "   s_sn_gt_1k:\t\t     f, t."                                         
[62] "   s_sn_gt_2k:\t\t     f, t."                                         
[63] "   s_sn_gt_4k:\t\t     f, t."                                         
[64] "   speech():\t\t     normal,good,very_good,very_poor,poor,unmeasured."
[65] "   static_normal:\t     f, t."                                        
[66] "   tymp():\t\t     a,as,b,ad,c."                                      
[67] "   viith_nerve_signs:        f, t."                                   
[68] "   wave_V_delayed:\t     f, t."                                       
[69] "   waveform_ItoV_prolonged:  f, t."                                   </code></pre>
</div>
</div>
<p>This line of code uses the str_split_fixed() function to split the names vector whenever it encounters a “:” character and the 2 argument indicates that it should split each element imto at most two parts.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> <span class="fu">str_split_fixed</span>(names, <span class="st">":"</span>, <span class="dv">2</span>) <span class="co">#split on regular expression pattern ":", this function returns a matrix</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      [,1]                        
 [1,] "   age_gt_60"              
 [2,] "   air()"                  
 [3,] "   airBoneGap"             
 [4,] "   ar_c()"                 
 [5,] "   ar_u()"                 
 [6,] "   bone()"                 
 [7,] "   boneAbnormal"           
 [8,] "   bser()"                 
 [9,] "   history_buzzing"        
[10,] "   history_dizziness"      
[11,] "   history_fluctuating"    
[12,] "   history_fullness"       
[13,] "   history_heredity"       
[14,] "   history_nausea"         
[15,] "   history_noise"          
[16,] "   history_recruitment"    
[17,] "   history_ringing"        
[18,] "   history_roaring"        
[19,] "   history_vomiting"       
[20,] "   late_wave_poor"         
[21,] "   m_at_2k"                
[22,] "   m_cond_lt_1k"           
[23,] "   m_gt_1k"                
[24,] "   m_m_gt_2k"              
[25,] "   m_m_sn"                 
[26,] "   m_m_sn_gt_1k"           
[27,] "   m_m_sn_gt_2k"           
[28,] "   m_m_sn_gt_500"          
[29,] "   m_p_sn_gt_2k"           
[30,] "   m_s_gt_500"             
[31,] "   m_s_sn"                 
[32,] "   m_s_sn_gt_1k"           
[33,] "   m_s_sn_gt_2k"           
[34,] "   m_s_sn_gt_3k"           
[35,] "   m_s_sn_gt_4k"           
[36,] "   m_sn_2_3k"              
[37,] "   m_sn_gt_1k"             
[38,] "   m_sn_gt_2k"             
[39,] "   m_sn_gt_3k"             
[40,] "   m_sn_gt_4k"             
[41,] "   m_sn_gt_500"            
[42,] "   m_sn_gt_6k"             
[43,] "   m_sn_lt_1k"             
[44,] "   m_sn_lt_2k"             
[45,] "   m_sn_lt_3k"             
[46,] "   middle_wave_poor"       
[47,] "   mod_gt_4k"              
[48,] "   mod_mixed"              
[49,] "   mod_s_mixed"            
[50,] "   mod_s_sn_gt_500"        
[51,] "   mod_sn"                 
[52,] "   mod_sn_gt_1k"           
[53,] "   mod_sn_gt_2k"           
[54,] "   mod_sn_gt_3k"           
[55,] "   mod_sn_gt_4k"           
[56,] "   mod_sn_gt_500"          
[57,] "   notch_4k"               
[58,] "   notch_at_4k"            
[59,] "   o_ar_c()"               
[60,] "   o_ar_u()"               
[61,] "   s_sn_gt_1k"             
[62,] "   s_sn_gt_2k"             
[63,] "   s_sn_gt_4k"             
[64,] "   speech()"               
[65,] "   static_normal"          
[66,] "   tymp()"                 
[67,] "   viith_nerve_signs"      
[68,] "   wave_V_delayed"         
[69,] "   waveform_ItoV_prolonged"
      [,2]                                                       
 [1,] "\t\t     f, t."                                           
 [2,] "\t\t     mild,moderate,severe,normal,profound."           
 [3,] "\t\t     f, t."                                           
 [4,] "\t\t     normal,elevated,absent."                         
 [5,] "\t\t     normal,absent,elevated."                         
 [6,] "\t\t     mild,moderate,normal,unmeasured."                
 [7,] "\t     f, t."                                             
 [8,] "\t\t     normal,degraded."                                
 [9,] "\t     f, t."                                             
[10,] "\t     f, t."                                             
[11,] "\t     f, t."                                             
[12,] "\t     f, t."                                             
[13,] "\t     f, t."                                             
[14,] "\t     f, t."                                             
[15,] "\t     f, t."                                             
[16,] "\t     f, t."                                             
[17,] "\t     f, t."                                             
[18,] "\t     f, t."                                             
[19,] "\t     f, t."                                             
[20,] "\t     f, t."                                             
[21,] "\t\t     f, t."                                           
[22,] "\t     f, t."                                             
[23,] "\t\t     f, t."                                           
[24,] "\t\t     f, t."                                           
[25,] "\t\t     f, t."                                           
[26,] "\t     f, t."                                             
[27,] "\t     f, t."                                             
[28,] "\t     f, t."                                             
[29,] "\t     f, t."                                             
[30,] "\t\t     f, t."                                           
[31,] "\t\t     f, t."                                           
[32,] "\t     f, t."                                             
[33,] "\t     f, t."                                             
[34,] "\t     f, t."                                             
[35,] "\t     f, t."                                             
[36,] "\t\t     f, t."                                           
[37,] "\t\t     f, t."                                           
[38,] "\t\t     f, t."                                           
[39,] "\t\t     f, t."                                           
[40,] "\t\t     f, t."                                           
[41,] "\t\t     f, t."                                           
[42,] "\t\t     f, t."                                           
[43,] "\t\t     f, t. "                                          
[44,] "\t\t     f, t."                                           
[45,] "\t\t     f, t."                                           
[46,] "\t     f, t."                                             
[47,] "\t\t     f, t."                                           
[48,] "\t\t     f, t."                                           
[49,] "\t\t     f, t."                                           
[50,] "\t     f, t."                                             
[51,] "\t\t     f, t."                                           
[52,] "\t     f, t."                                             
[53,] "\t     f, t."                                             
[54,] "\t     f, t."                                             
[55,] "\t     f, t."                                             
[56,] "\t     f, t."                                             
[57,] "\t\t     f, t."                                           
[58,] "\t\t     f, t."                                           
[59,] "\t\t     normal,elevated,absent."                         
[60,] "\t\t     normal,absent,elevated."                         
[61,] "\t\t     f, t."                                           
[62,] "\t\t     f, t."                                           
[63,] "\t\t     f, t."                                           
[64,] "\t\t     normal,good,very_good,very_poor,poor,unmeasured."
[65,] "\t     f, t."                                             
[66,] "\t\t     a,as,b,ad,c."                                    
[67,] "        f, t."                                            
[68,] "\t     f, t."                                             
[69,] "  f, t."                                                  </code></pre>
</div>
</div>
<p>The below code selects the first column and it’s set of values and displays it</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span> names[,<span class="dv">1</span>]</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "   age_gt_60"               "   air()"                  
 [3] "   airBoneGap"              "   ar_c()"                 
 [5] "   ar_u()"                  "   bone()"                 
 [7] "   boneAbnormal"            "   bser()"                 
 [9] "   history_buzzing"         "   history_dizziness"      
[11] "   history_fluctuating"     "   history_fullness"       
[13] "   history_heredity"        "   history_nausea"         
[15] "   history_noise"           "   history_recruitment"    
[17] "   history_ringing"         "   history_roaring"        
[19] "   history_vomiting"        "   late_wave_poor"         
[21] "   m_at_2k"                 "   m_cond_lt_1k"           
[23] "   m_gt_1k"                 "   m_m_gt_2k"              
[25] "   m_m_sn"                  "   m_m_sn_gt_1k"           
[27] "   m_m_sn_gt_2k"            "   m_m_sn_gt_500"          
[29] "   m_p_sn_gt_2k"            "   m_s_gt_500"             
[31] "   m_s_sn"                  "   m_s_sn_gt_1k"           
[33] "   m_s_sn_gt_2k"            "   m_s_sn_gt_3k"           
[35] "   m_s_sn_gt_4k"            "   m_sn_2_3k"              
[37] "   m_sn_gt_1k"              "   m_sn_gt_2k"             
[39] "   m_sn_gt_3k"              "   m_sn_gt_4k"             
[41] "   m_sn_gt_500"             "   m_sn_gt_6k"             
[43] "   m_sn_lt_1k"              "   m_sn_lt_2k"             
[45] "   m_sn_lt_3k"              "   middle_wave_poor"       
[47] "   mod_gt_4k"               "   mod_mixed"              
[49] "   mod_s_mixed"             "   mod_s_sn_gt_500"        
[51] "   mod_sn"                  "   mod_sn_gt_1k"           
[53] "   mod_sn_gt_2k"            "   mod_sn_gt_3k"           
[55] "   mod_sn_gt_4k"            "   mod_sn_gt_500"          
[57] "   notch_4k"                "   notch_at_4k"            
[59] "   o_ar_c()"                "   o_ar_u()"               
[61] "   s_sn_gt_1k"              "   s_sn_gt_2k"             
[63] "   s_sn_gt_4k"              "   speech()"               
[65] "   static_normal"           "   tymp()"                 
[67] "   viith_nerve_signs"       "   wave_V_delayed"         
[69] "   waveform_ItoV_prolonged"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>names <span class="ot">&lt;-</span><span class="fu">str_trim</span>(names) <span class="sc">|&gt;</span> <span class="fu">str_replace_all</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">(|</span><span class="sc">\\</span><span class="st">)"</span>, <span class="st">""</span>) <span class="co"># we use a pipe, and another reg exp "\\(|\\)", \\ is the escape.</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "age_gt_60"               "air"                    
 [3] "airBoneGap"              "ar_c"                   
 [5] "ar_u"                    "bone"                   
 [7] "boneAbnormal"            "bser"                   
 [9] "history_buzzing"         "history_dizziness"      
[11] "history_fluctuating"     "history_fullness"       
[13] "history_heredity"        "history_nausea"         
[15] "history_noise"           "history_recruitment"    
[17] "history_ringing"         "history_roaring"        
[19] "history_vomiting"        "late_wave_poor"         
[21] "m_at_2k"                 "m_cond_lt_1k"           
[23] "m_gt_1k"                 "m_m_gt_2k"              
[25] "m_m_sn"                  "m_m_sn_gt_1k"           
[27] "m_m_sn_gt_2k"            "m_m_sn_gt_500"          
[29] "m_p_sn_gt_2k"            "m_s_gt_500"             
[31] "m_s_sn"                  "m_s_sn_gt_1k"           
[33] "m_s_sn_gt_2k"            "m_s_sn_gt_3k"           
[35] "m_s_sn_gt_4k"            "m_sn_2_3k"              
[37] "m_sn_gt_1k"              "m_sn_gt_2k"             
[39] "m_sn_gt_3k"              "m_sn_gt_4k"             
[41] "m_sn_gt_500"             "m_sn_gt_6k"             
[43] "m_sn_lt_1k"              "m_sn_lt_2k"             
[45] "m_sn_lt_3k"              "middle_wave_poor"       
[47] "mod_gt_4k"               "mod_mixed"              
[49] "mod_s_mixed"             "mod_s_sn_gt_500"        
[51] "mod_sn"                  "mod_sn_gt_1k"           
[53] "mod_sn_gt_2k"            "mod_sn_gt_3k"           
[55] "mod_sn_gt_4k"            "mod_sn_gt_500"          
[57] "notch_4k"                "notch_at_4k"            
[59] "o_ar_c"                  "o_ar_u"                 
[61] "s_sn_gt_1k"              "s_sn_gt_2k"             
[63] "s_sn_gt_4k"              "speech"                 
[65] "static_normal"           "tymp"                   
[67] "viith_nerve_signs"       "wave_V_delayed"         
[69] "waveform_ItoV_prolonged"</code></pre>
</div>
</div>
<p>The below code sets the names to the columns from 1 to 69. It takes the first 69 columns and assigns the corresponding values from the “names” vector.</p>
<p>This is a common step in data analysis to ensure that the columns of the data frame have meaningful names associated with the variables they represent.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(data)[<span class="dv">1</span><span class="sc">:</span><span class="dv">69</span>] <span class="ot">&lt;-</span> names</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 200 × 71
   age_gt_60 air      airBoneGap ar_c     ar_u   bone  boneAbnormal bser 
   &lt;lgl&gt;     &lt;chr&gt;    &lt;lgl&gt;      &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt; &lt;lgl&gt;        &lt;chr&gt;
 1 FALSE     mild     FALSE      normal   normal &lt;NA&gt;  TRUE         &lt;NA&gt; 
 2 FALSE     moderate FALSE      normal   normal &lt;NA&gt;  TRUE         &lt;NA&gt; 
 3 TRUE      mild     TRUE       &lt;NA&gt;     absent mild  TRUE         &lt;NA&gt; 
 4 TRUE      mild     TRUE       &lt;NA&gt;     absent mild  FALSE        &lt;NA&gt; 
 5 TRUE      mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 6 TRUE      mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 7 FALSE     mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 8 FALSE     mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 9 FALSE     severe   FALSE      &lt;NA&gt;     &lt;NA&gt;   &lt;NA&gt;  TRUE         &lt;NA&gt; 
10 TRUE      mild     FALSE      elevated absent mild  TRUE         &lt;NA&gt; 
# ℹ 190 more rows
# ℹ 63 more variables: history_buzzing &lt;lgl&gt;, history_dizziness &lt;lgl&gt;,
#   history_fluctuating &lt;lgl&gt;, history_fullness &lt;lgl&gt;, history_heredity &lt;lgl&gt;,
#   history_nausea &lt;lgl&gt;, history_noise &lt;lgl&gt;, history_recruitment &lt;lgl&gt;,
#   history_ringing &lt;lgl&gt;, history_roaring &lt;lgl&gt;, history_vomiting &lt;lgl&gt;,
#   late_wave_poor &lt;lgl&gt;, m_at_2k &lt;lgl&gt;, m_cond_lt_1k &lt;lgl&gt;, m_gt_1k &lt;lgl&gt;,
#   m_m_gt_2k &lt;lgl&gt;, m_m_sn &lt;lgl&gt;, m_m_sn_gt_1k &lt;lgl&gt;, m_m_sn_gt_2k &lt;lgl&gt;, …</code></pre>
</div>
</div>
<p>The below line of code assigns the column name “id” to column 70 and name “class” to column 71.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(data)[<span class="dv">70</span><span class="sc">:</span><span class="dv">71</span>] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"id"</span>, <span class="st">"class"</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 200 × 71
   age_gt_60 air      airBoneGap ar_c     ar_u   bone  boneAbnormal bser 
   &lt;lgl&gt;     &lt;chr&gt;    &lt;lgl&gt;      &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt; &lt;lgl&gt;        &lt;chr&gt;
 1 FALSE     mild     FALSE      normal   normal &lt;NA&gt;  TRUE         &lt;NA&gt; 
 2 FALSE     moderate FALSE      normal   normal &lt;NA&gt;  TRUE         &lt;NA&gt; 
 3 TRUE      mild     TRUE       &lt;NA&gt;     absent mild  TRUE         &lt;NA&gt; 
 4 TRUE      mild     TRUE       &lt;NA&gt;     absent mild  FALSE        &lt;NA&gt; 
 5 TRUE      mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 6 TRUE      mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 7 FALSE     mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 8 FALSE     mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 9 FALSE     severe   FALSE      &lt;NA&gt;     &lt;NA&gt;   &lt;NA&gt;  TRUE         &lt;NA&gt; 
10 TRUE      mild     FALSE      elevated absent mild  TRUE         &lt;NA&gt; 
# ℹ 190 more rows
# ℹ 63 more variables: history_buzzing &lt;lgl&gt;, history_dizziness &lt;lgl&gt;,
#   history_fluctuating &lt;lgl&gt;, history_fullness &lt;lgl&gt;, history_heredity &lt;lgl&gt;,
#   history_nausea &lt;lgl&gt;, history_noise &lt;lgl&gt;, history_recruitment &lt;lgl&gt;,
#   history_ringing &lt;lgl&gt;, history_roaring &lt;lgl&gt;, history_vomiting &lt;lgl&gt;,
#   late_wave_poor &lt;lgl&gt;, m_at_2k &lt;lgl&gt;, m_cond_lt_1k &lt;lgl&gt;, m_gt_1k &lt;lgl&gt;,
#   m_m_gt_2k &lt;lgl&gt;, m_m_sn &lt;lgl&gt;, m_m_sn_gt_1k &lt;lgl&gt;, m_m_sn_gt_2k &lt;lgl&gt;, …</code></pre>
</div>
</div>
<p>The below code helps filter out the rows with one or more missing values using the filter() and !complete.cases() function. These rows can further be handled through removal, imputation or other methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>missing.value.rows <span class="ot">&lt;-</span> data <span class="sc">|&gt;</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">complete.cases</span>(data))</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>missing.value.rows</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 196 × 71
   age_gt_60 air      airBoneGap ar_c     ar_u   bone  boneAbnormal bser 
   &lt;lgl&gt;     &lt;chr&gt;    &lt;lgl&gt;      &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt; &lt;lgl&gt;        &lt;chr&gt;
 1 FALSE     mild     FALSE      normal   normal &lt;NA&gt;  TRUE         &lt;NA&gt; 
 2 FALSE     moderate FALSE      normal   normal &lt;NA&gt;  TRUE         &lt;NA&gt; 
 3 TRUE      mild     TRUE       &lt;NA&gt;     absent mild  TRUE         &lt;NA&gt; 
 4 TRUE      mild     TRUE       &lt;NA&gt;     absent mild  FALSE        &lt;NA&gt; 
 5 TRUE      mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 6 TRUE      mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 7 FALSE     mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 8 FALSE     mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 9 FALSE     severe   FALSE      &lt;NA&gt;     &lt;NA&gt;   &lt;NA&gt;  TRUE         &lt;NA&gt; 
10 TRUE      mild     FALSE      elevated absent mild  TRUE         &lt;NA&gt; 
# ℹ 186 more rows
# ℹ 63 more variables: history_buzzing &lt;lgl&gt;, history_dizziness &lt;lgl&gt;,
#   history_fluctuating &lt;lgl&gt;, history_fullness &lt;lgl&gt;, history_heredity &lt;lgl&gt;,
#   history_nausea &lt;lgl&gt;, history_noise &lt;lgl&gt;, history_recruitment &lt;lgl&gt;,
#   history_ringing &lt;lgl&gt;, history_roaring &lt;lgl&gt;, history_vomiting &lt;lgl&gt;,
#   late_wave_poor &lt;lgl&gt;, m_at_2k &lt;lgl&gt;, m_cond_lt_1k &lt;lgl&gt;, m_gt_1k &lt;lgl&gt;,
#   m_m_gt_2k &lt;lgl&gt;, m_m_sn &lt;lgl&gt;, m_m_sn_gt_1k &lt;lgl&gt;, m_m_sn_gt_2k &lt;lgl&gt;, …</code></pre>
</div>
</div>
<p>The below line of code creates a new column with the help of the mutate() function with the name “na_count”. This new column calculates the sum of NA values in each row. This helps in understanding data quality in each row.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">na_count =</span> <span class="fu">rowSums</span>(<span class="fu">is.na</span>(data)))</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 200 × 72
   age_gt_60 air      airBoneGap ar_c     ar_u   bone  boneAbnormal bser 
   &lt;lgl&gt;     &lt;chr&gt;    &lt;lgl&gt;      &lt;chr&gt;    &lt;chr&gt;  &lt;chr&gt; &lt;lgl&gt;        &lt;chr&gt;
 1 FALSE     mild     FALSE      normal   normal &lt;NA&gt;  TRUE         &lt;NA&gt; 
 2 FALSE     moderate FALSE      normal   normal &lt;NA&gt;  TRUE         &lt;NA&gt; 
 3 TRUE      mild     TRUE       &lt;NA&gt;     absent mild  TRUE         &lt;NA&gt; 
 4 TRUE      mild     TRUE       &lt;NA&gt;     absent mild  FALSE        &lt;NA&gt; 
 5 TRUE      mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 6 TRUE      mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 7 FALSE     mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 8 FALSE     mild     FALSE      normal   normal mild  TRUE         &lt;NA&gt; 
 9 FALSE     severe   FALSE      &lt;NA&gt;     &lt;NA&gt;   &lt;NA&gt;  TRUE         &lt;NA&gt; 
10 TRUE      mild     FALSE      elevated absent mild  TRUE         &lt;NA&gt; 
# ℹ 190 more rows
# ℹ 64 more variables: history_buzzing &lt;lgl&gt;, history_dizziness &lt;lgl&gt;,
#   history_fluctuating &lt;lgl&gt;, history_fullness &lt;lgl&gt;, history_heredity &lt;lgl&gt;,
#   history_nausea &lt;lgl&gt;, history_noise &lt;lgl&gt;, history_recruitment &lt;lgl&gt;,
#   history_ringing &lt;lgl&gt;, history_roaring &lt;lgl&gt;, history_vomiting &lt;lgl&gt;,
#   late_wave_poor &lt;lgl&gt;, m_at_2k &lt;lgl&gt;, m_cond_lt_1k &lt;lgl&gt;, m_gt_1k &lt;lgl&gt;,
#   m_m_gt_2k &lt;lgl&gt;, m_m_sn &lt;lgl&gt;, m_m_sn_gt_1k &lt;lgl&gt;, m_m_sn_gt_2k &lt;lgl&gt;, …</code></pre>
</div>
</div>
<p>The below code calculates the sum of NA values in each column of the “data” data frame through the -sum(is,na(.)) part of code. The summarize() function summarizes the output and creates a new data frame.</p>
<p>The pivot_longer() function changes the output from a wide to a long format and columns “column_names” and “na_count” are added to the output. Finally the arrange() function arranges data in the ascending value of NA count.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>data <span class="sc">|&gt;</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="fu">across</span>(<span class="fu">everything</span>(), <span class="sc">~</span><span class="fu">sum</span>(<span class="fu">is.na</span>(.)), <span class="at">.names =</span> <span class="st">"na_{.col}"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">everything</span>(), <span class="at">names_to =</span> <span class="st">"column_name"</span>, <span class="at">values_to =</span> <span class="st">"na_count"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(na_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 72 × 2
   column_name            na_count
   &lt;chr&gt;                     &lt;int&gt;
 1 na_age_gt_60                  0
 2 na_air                        0
 3 na_airBoneGap                 0
 4 na_boneAbnormal               0
 5 na_history_buzzing            0
 6 na_history_dizziness          0
 7 na_history_fluctuating        0
 8 na_history_fullness           0
 9 na_history_heredity           0
10 na_history_nausea             0
# ℹ 62 more rows</code></pre>
</div>
</div>
<p>The below code does the same as the above one expect that row number 8 is removed from the data frame using the select() function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>data.bser.removed <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="dv">8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(<span class="fu">everything</span>(), <span class="sc">~</span><span class="fu">sum</span>(<span class="fu">is.na</span>(.)), <span class="at">.names =</span> <span class="st">"na_{.col}"</span>))</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>data.bser.removed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 71
  na_age_gt_60 na_air na_airBoneGap na_ar_c na_ar_u na_bone na_boneAbnormal
         &lt;int&gt;  &lt;int&gt;         &lt;int&gt;   &lt;int&gt;   &lt;int&gt;   &lt;int&gt;           &lt;int&gt;
1            0      0             0       4       3      75               0
# ℹ 64 more variables: na_history_buzzing &lt;int&gt;, na_history_dizziness &lt;int&gt;,
#   na_history_fluctuating &lt;int&gt;, na_history_fullness &lt;int&gt;,
#   na_history_heredity &lt;int&gt;, na_history_nausea &lt;int&gt;, na_history_noise &lt;int&gt;,
#   na_history_recruitment &lt;int&gt;, na_history_ringing &lt;int&gt;,
#   na_history_roaring &lt;int&gt;, na_history_vomiting &lt;int&gt;,
#   na_late_wave_poor &lt;int&gt;, na_m_at_2k &lt;int&gt;, na_m_cond_lt_1k &lt;int&gt;,
#   na_m_gt_1k &lt;int&gt;, na_m_m_gt_2k &lt;int&gt;, na_m_m_sn &lt;int&gt;, …</code></pre>
</div>
</div>
<p>The data data frame in the below code will retain only the columns that do not have “bser” in their names, effectively removing columns with names containing that term.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> data <span class="sc">%&gt;%</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span><span class="fu">matches</span>(<span class="st">"bser"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The mistaken variable in the code is a numeric vector that contains the values 2, 3, 4, and “?”. However, because one of the values is a character (the “?”), R will coerce all the values in the vector to characters, making it a character vector.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>mistaken <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="st">"?"</span>)</span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(mistaken)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
</div>
<p>The mistaken character vector is fixed by converting it into an integer vector using the parse_integer() function from the readr package. Additionally, it specifies that the “?” character should be treated as NA (missing value).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>fixed <span class="ot">&lt;-</span> <span class="fu">parse_integer</span>(mistaken, <span class="at">na =</span> <span class="st">'?'</span>)</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>fixed</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  2  3  4 NA</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(fixed)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "integer"</code></pre>
</div>
</div>
<p>This code should load the “algae” dataset and display the 48th row of data from the “algae” dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"DMwR2"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into 'C:/Users/Joel Jojo/AppData/Local/R/win-library/4.3'
(as 'lib' is unspecified)</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in contrib.url(repos, "source"): trying to use CRAN without setting a mirror</code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DMwR2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 method overwritten by 'quantmod':
  method            from
  as.zoo.data.frame zoo </code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(algae, <span class="at">package =</span> <span class="st">"DMwR2"</span>)</span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a>algae[<span class="dv">48</span>,]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 18
  season size  speed  mxPH  mnO2    Cl   NO3   NH4  oPO4   PO4  Chla    a1    a2
  &lt;fct&gt;  &lt;fct&gt; &lt;fct&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 winter small low      NA  12.6     9  0.23    10     5     6   1.1  35.5     0
# ℹ 5 more variables: a3 &lt;dbl&gt;, a4 &lt;dbl&gt;, a5 &lt;dbl&gt;, a6 &lt;dbl&gt;, a7 &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The below code initializes a ggplot2 using the “algae” dataset using the mxPH variable as the sample data. The geom_qq_band() function adds a shaded region representing the confidence intervals. The stat_qq_point() function adds points to the qq plots to represent the observed quantiles of the mxPH variables. This plot allows to visually assess wether the data follows a normal distribution. The reference line helps you identify deviations from normality.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot a QQ plot of mxPH</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"car"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into 'C:/Users/Joel Jojo/AppData/Local/R/win-library/4.3'
(as 'lib' is unspecified)</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in contrib.url(repos, "source"): trying to use CRAN without setting a mirror</code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(car)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: carData</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'car'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:purrr':

    some</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    recode</code></pre>
</div>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(algae, <span class="fu">aes</span>(<span class="at">sample =</span> mxPH)) <span class="sc">+</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_band</span>() <span class="sc">+</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_qq_point</span>() <span class="sc">+</span></span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_qq_line</span>(<span class="at">color =</span> <span class="st">"red"</span>, <span class="at">method =</span> <span class="st">"identity"</span>, <span class="at">intercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">slope =</span> <span class="dv">1</span>) <span class="sc">+</span>  </span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Normal QQ plot of mxPH"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in stat_qq_line(color = "red", method = "identity", intercept = -2, :
Ignoring unknown parameters: `method`, `intercept`, and `slope`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Exercise-2_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The below line of code imputes any missing values from the 48th row with the mean of the mxPH variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>algae <span class="ot">&lt;-</span> algae <span class="sc">|&gt;</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">mxPH =</span> <span class="fu">ifelse</span>(<span class="fu">row_number</span>() <span class="sc">==</span> <span class="dv">48</span>, <span class="fu">mean</span>(mxPH, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), mxPH))</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>algae</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 200 × 18
   season size  speed   mxPH  mnO2    Cl    NO3   NH4  oPO4   PO4  Chla    a1
   &lt;fct&gt;  &lt;fct&gt; &lt;fct&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 winter small medium  8      9.8  60.8  6.24  578   105   170   50      0  
 2 spring small medium  8.35   8    57.8  1.29  370   429.  559.   1.3    1.4
 3 autumn small medium  8.1   11.4  40.0  5.33  347.  126.  187.  15.6    3.3
 4 spring small medium  8.07   4.8  77.4  2.30   98.2  61.2 139.   1.4    3.1
 5 autumn small medium  8.06   9    55.4 10.4   234.   58.2  97.6 10.5    9.2
 6 winter small high    8.25  13.1  65.8  9.25  430    18.2  56.7 28.4   15.1
 7 summer small high    8.15  10.3  73.2  1.54  110    61.2 112.   3.2    2.4
 8 autumn small high    8.05  10.6  59.1  4.99  206.   44.7  77.4  6.9   18.2
 9 winter small medium  8.7    3.4  22.0  0.886 103.   36.3  71    5.54  25.4
10 winter small high    7.93   9.9   8    1.39    5.8  27.2  46.6  0.8   17  
# ℹ 190 more rows
# ℹ 6 more variables: a2 &lt;dbl&gt;, a3 &lt;dbl&gt;, a4 &lt;dbl&gt;, a5 &lt;dbl&gt;, a6 &lt;dbl&gt;,
#   a7 &lt;dbl&gt;</code></pre>
</div>
</div>
<p>The below code plots the same QQ plot as before but with the variable “Chla”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(algae, <span class="fu">aes</span>(<span class="at">sample =</span> Chla)) <span class="sc">+</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_qq_band</span>() <span class="sc">+</span></span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">stat_qq_point</span>() <span class="sc">+</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_qq_line</span>(<span class="at">color =</span> <span class="st">"red"</span>, <span class="at">method =</span> <span class="st">"identity"</span>, <span class="at">intercept =</span> <span class="sc">-</span><span class="dv">2</span>, <span class="at">slope =</span> <span class="dv">1</span>) <span class="sc">+</span>  </span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Normal QQ plot of Chla"</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in stat_qq_line(color = "red", method = "identity", intercept = -2, :
Ignoring unknown parameters: `method`, `intercept`, and `slope`</code></pre>
</div>
<div class="cell-output-display">
<p><img src="Exercise-2_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The below code gives the median value of the “Chla” variable in the dataset, which provides a measure of the central tendency of the non-missing values in that variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">median</span>(algae<span class="sc">$</span>Chla, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 5.475</code></pre>
</div>
</div>
<p>The below code provides the mean value of the “Chla” variable in the dataset, based on the non-missing values. It provides a measure of the central tendency of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(algae<span class="sc">$</span>Chla, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 13.9712</code></pre>
</div>
</div>
<p>The below code removes missing values from Chla variable and replaces it with the median of the non missing values from the Chla column. This common technique for imputing missing data helps make the dataset more complete for analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a>algae <span class="ot">&lt;-</span> algae <span class="sc">|&gt;</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Chla =</span> <span class="fu">if_else</span>(<span class="fu">is.na</span>(Chla), <span class="fu">median</span>(Chla, <span class="at">na.rm =</span> <span class="cn">TRUE</span>), Chla))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The below code assigns the columns 4 to 18 to the to the variable algae_numeric and then the NA values are removed using the drop_na() function. Finally a correlation matrix of the numeric variables in algae_numeric using correlate() function. This is finally plotted using the plot() function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>algae_numeric <span class="ot">&lt;-</span> algae[, <span class="dv">4</span><span class="sc">:</span><span class="dv">18</span>] <span class="sc">%&gt;%</span></span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>()  <span class="co"># Removes rows with NA values</span></span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb98-4"><a href="#cb98-4" aria-hidden="true" tabindex="-1"></a>cor_matrix <span class="ot">&lt;-</span> algae_numeric <span class="sc">|&gt;</span> <span class="fu">correlate</span>() <span class="sc">|&gt;</span> <span class="fu">plot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Exercise-2_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a>cor_matrix</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Exercise-2_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The below code filters out rows where the missing values percentage exceeds 20%. This is then used in a linear regression model where the dependant variable is PO4 and the independant variable is oPO4.</p>
<p>lm(formula = PO4 ~ oPO4, data = algae) is the linear regression model formula used.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a>algae <span class="ot">&lt;-</span> algae <span class="sc">%&gt;%</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">rowSums</span>(<span class="fu">is.na</span>(.)) <span class="sc">/</span> <span class="fu">ncol</span>(.) <span class="sc">&lt;</span> <span class="fl">0.2</span>)<span class="co">#this is a method provided that selects the observations with 20% or move values as NAs. </span></span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>m <span class="ot">=</span> <span class="fu">lm</span>(PO4 <span class="sc">~</span> oPO4, <span class="at">data =</span> algae)</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a><span class="fu">lm</span>(<span class="at">formula =</span> PO4 <span class="sc">~</span> oPO4, <span class="at">data =</span> algae)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = PO4 ~ oPO4, data = algae)

Coefficients:
(Intercept)         oPO4  
     42.897        1.293  </code></pre>
</div>
</div>
<p>The summary() function here provides a desciprtion of the linear regressoin model object, including coefficients, standard errors and other predictors that describes the model’s fit.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a>m <span class="sc">|&gt;</span> </span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = PO4 ~ oPO4, data = algae)

Residuals:
    Min      1Q  Median      3Q     Max 
-110.12  -36.34  -12.68   23.26  216.98 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)   42.897      4.808   8.922 3.34e-16 ***
oPO4           1.293      0.041  31.535  &lt; 2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 52.37 on 195 degrees of freedom
  (1 observation deleted due to missingness)
Multiple R-squared:  0.8361,    Adjusted R-squared:  0.8352 
F-statistic: 994.5 on 1 and 195 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>The tidy() function represents the linear regerssion object model in a tabular and organized manner, making it easier to work with the results of the regression model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>m <span class="sc">|&gt;</span> </span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summary</span>() <span class="sc">|&gt;</span> </span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  term        estimate std.error statistic  p.value
  &lt;chr&gt;          &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)    42.9     4.81        8.92 3.34e-16
2 oPO4            1.29    0.0410     31.5  1.68e-78</code></pre>
</div>
</div>
<p>By using algae$PO4, we retrieve and access the values of the “PO4” variable from the “algae” dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>algae<span class="sc">$</span>PO4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] 170.000 558.750 187.057 138.700  97.580  56.667 111.750  77.434  71.000
 [10]  46.600  20.750  19.000  17.000  15.000  61.600  98.250  50.000  57.833
 [19]  61.500 771.600 586.000  18.000  40.000  27.500  11.500  44.136  13.600
 [28]      NA  45.000  19.000 142.000 304.000 130.750  47.000  23.000  84.460
 [37]   3.000   3.000 253.250 255.280 296.000 175.046 344.600 326.857  40.667
 [46]  43.500  39.000   6.000 121.000  20.812  49.333  22.900  11.800  11.818
 [55]   6.500   1.000   4.000   6.000  11.000   6.000  14.000   6.667   6.750
 [64]   7.200   6.000  10.750   2.500 351.600 313.600 279.066 152.333  58.623
 [73] 249.250 233.500 215.500 102.333 105.727 111.375 108.000  56.667  60.000
 [82] 104.000  69.930 214.000 254.600 169.001 607.167 624.733 303.333 391.750
 [91] 265.250 232.833 244.000 218.000 138.500 239.000 235.667 205.875 211.667
[100] 186.500 154.125 183.667 292.625 285.714 201.778 275.143 124.200 141.833
[109] 132.546  17.333  26.000  16.662 102.571  86.997  10.111  18.293  13.200
[118] 432.909 320.400 287.000 262.727 222.286 122.000 127.222  89.625 284.000
[127] 277.333 177.625  72.900  82.444  66.750 173.750 317.000  84.000 213.000
[136]  88.500 115.000  98.143 143.750 197.143  35.200  23.485 200.231 147.833
[145] 276.000 123.333  75.207 116.200 188.667  72.696 116.200  34.000  76.333
[154]  58.374 361.000 236.000 125.800  54.916  75.333 186.000 252.500 269.667
[163]  46.438  85.000 171.272 232.900 146.452 246.667 219.909 272.222 388.167
[172] 167.900 137.778 194.100 221.278  21.300  11.000  14.354   7.000   6.200
[181]   7.654  16.000  56.091  52.875 228.364  85.400  87.125 101.455 127.000
[190]  81.558  50.455 120.889  91.111  61.444  79.750  75.904 140.220 140.517</code></pre>
</div>
</div>
<p>Here, the PO4 variable is mutated with the help of the mutate function. The 28th row in the PO4 variable is replaced with a new value based on arithmetic operation on the oPO4 variable. This is a way to conditionally update a specific row in a data frame based on a row number.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>algae <span class="ot">&lt;-</span> algae <span class="sc">%&gt;%</span></span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">PO4 =</span> <span class="fu">ifelse</span>(<span class="fu">row_number</span>() <span class="sc">==</span> <span class="dv">28</span>, <span class="fl">42.897</span> <span class="sc">+</span> <span class="fl">1.293</span> <span class="sc">*</span> oPO4, PO4))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>the resid() function here calculates the residuals of the linear regression model m. Residuals represent the differences between the observed values and the values predicted by the model.</p>
<p>algae %&gt;% filter(row_number() != 28): it creates a reduced dataset without the 28th row.</p>
<p>The pull() function is used to extract the values of the “oPO4” variable from the filtered dataset and assigns them to the oPO4_reduced variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a>res <span class="ot">=</span> <span class="fu">resid</span>(m)</span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a>oPO4_reduced <span class="ot">&lt;-</span> algae <span class="sc">%&gt;%</span></span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">row_number</span>() <span class="sc">!=</span> <span class="dv">28</span>) <span class="sc">%&gt;%</span></span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(oPO4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The below code utilizes a ggplot to portray a residual plot, which is a tool for assessing the assumptions of a linear regression model. The aes() function specifies that “oPO4” (predictor variable) should be plotted on the x-axis and “res” (containing residuals) should be plotted on the y-axis.</p>
<p>This plot allows to visually inspect whether the residuals are randomly distributed around zero, which is a key assumption for linear regression</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> <span class="fu">data.frame</span>(<span class="at">oPO4 =</span> m<span class="sc">$</span>model<span class="sc">$</span>oPO4, <span class="at">res =</span> res), <span class="fu">aes</span>(<span class="at">x =</span> oPO4, <span class="at">y =</span> res)) <span class="sc">+</span></span>
<span id="cb110-2"><a href="#cb110-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb110-3"><a href="#cb110-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb110-4"><a href="#cb110-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(</span>
<span id="cb110-5"><a href="#cb110-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">x =</span> <span class="st">"oPO4"</span>,</span>
<span id="cb110-6"><a href="#cb110-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">y =</span> <span class="st">"residuals"</span>,</span>
<span id="cb110-7"><a href="#cb110-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">title =</span> <span class="st">"Residual Plot"</span></span>
<span id="cb110-8"><a href="#cb110-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Exercise-2_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>When the function fillPO4(x) is called, it will return a modified version of the input vector x where missing values have been replaced with values calculated using the formula 42.897 + 1.293 * x. This function is useful for imputing missing values in numeric variables based on a specified formula or calculation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a>fillPO4 <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">if_else</span>(<span class="fu">is.na</span>(x), <span class="fl">42.897</span> <span class="sc">+</span> <span class="fl">1.293</span> <span class="sc">*</span> x, x)</span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a><span class="co">#if x is not NA, return 42.897+1.293*x </span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the below code the sapply() function applies the previously created function fillPO4 to the values in the oPO4 column of the selected rows(where “PO4” is missing)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a>algae[<span class="fu">is.na</span>(algae<span class="sc">$</span>PO4), <span class="st">"PO4"</span>] <span class="ot">&lt;-</span> <span class="fu">sapply</span>(algae[<span class="fu">is.na</span>(algae<span class="sc">$</span>PO4), <span class="st">"oPO4"</span>], fillPO4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The function manyNAs() from the “DMwR2” package helps identify which rows have a significant amount of missing data. After running this code the “algae” dataset will be modified to exclude rows with a substantial amount of missing data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(algae, <span class="at">package=</span><span class="st">"DMwR2"</span>)</span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>algae <span class="ot">&lt;-</span> algae[<span class="sc">-</span><span class="fu">manyNAs</span>(algae), ] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The below code mainly utilizes the knnImputation() function which takes the weighted average of k (k=10) most similar samples to impute missing values. Then the dataset with missing values are reloaded again and the rows with many missing values are filtered out.<br>
Finally the knnImputation() function is used to impute missing values with the median of the k (k=10) most similar samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a>algae <span class="ot">&lt;-</span> <span class="fu">knnImputation</span>(algae, <span class="at">k =</span> <span class="dv">10</span>) <span class="co">#use the weighted average of k most similar samples</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-3"><a href="#cb114-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb114-4"><a href="#cb114-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(algae, <span class="at">package=</span><span class="st">"DMwR2"</span>) <span class="co">#get data again so there are unknown values</span></span>
<span id="cb114-5"><a href="#cb114-5" aria-hidden="true" tabindex="-1"></a>algae <span class="ot">&lt;-</span> algae[<span class="sc">-</span><span class="fu">manyNAs</span>(algae), ] </span>
<span id="cb114-6"><a href="#cb114-6" aria-hidden="true" tabindex="-1"></a>algae <span class="ot">&lt;-</span> <span class="fu">knnImputation</span>(algae, <span class="at">k =</span> <span class="dv">10</span>, <span class="at">meth=</span><span class="st">"median"</span>) <span class="co">#use the median of k most similar samples</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, the getAnywhere() function displays the source code of the knnImputation() function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb115"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb115-1"><a href="#cb115-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getAnywhere</span>(<span class="fu">knnImputation</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>A single object matching 'knnImputation' was found
It was found in the following places
  package:DMwR2
  namespace:DMwR2
with value

function (data, k = 10, scale = TRUE, meth = "weighAvg", distData = NULL) 
{
    n &lt;- nrow(data)
    if (!is.null(distData)) {
        distInit &lt;- n + 1
        data &lt;- rbind(data, distData)
    }
    else distInit &lt;- 1
    N &lt;- nrow(data)
    ncol &lt;- ncol(data)
    contAttrs &lt;- which(vapply(data, dplyr::type_sum, character(1)) %in% 
        c("dbl", "int"))
    nomAttrs &lt;- setdiff(seq.int(ncol), contAttrs)
    hasNom &lt;- length(nomAttrs)
    dm &lt;- data
    if (scale) 
        dm[, contAttrs] &lt;- scale(dm[, contAttrs])
    if (hasNom) 
        for (i in nomAttrs) dm[[i]] &lt;- as.integer(dm[[i]])
    dm &lt;- as.matrix(dm)
    nas &lt;- which(!complete.cases(dm))
    if (!is.null(distData)) 
        tgt.nas &lt;- nas[nas &lt;= n]
    else tgt.nas &lt;- nas
    if (length(tgt.nas) == 0) 
        warning("No case has missing values. Stopping as there is nothing to do.")
    xcomplete &lt;- dm[setdiff(distInit:N, nas), ]
    if (nrow(xcomplete) &lt; k) 
        stop("Not sufficient complete cases for computing neighbors.")
    for (i in tgt.nas) {
        tgtAs &lt;- which(is.na(dm[i, ]))
        dist &lt;- scale(xcomplete, dm[i, ], FALSE)
        xnom &lt;- setdiff(nomAttrs, tgtAs)
        if (length(xnom)) 
            dist[, xnom] &lt;- ifelse(dist[, xnom] &gt; 0, 1, dist[, 
                xnom])
        dist &lt;- dist[, -tgtAs]
        dist &lt;- sqrt(drop(dist^2 %*% rep(1, ncol(dist))))
        ks &lt;- order(dist)[seq(k)]
        for (j in tgtAs) if (meth == "median") 
            data[i, j] &lt;- centralValue(data[setdiff(distInit:N, 
                nas)[ks], j])
        else data[i, j] &lt;- centralValue(data[setdiff(distInit:N, 
            nas)[ks], j], exp(-dist[ks]))
    }
    data[1:n, ]
}
&lt;bytecode: 0x000002183550f468&gt;
&lt;environment: namespace:DMwR2&gt;</code></pre>
</div>
</div>
</section>
<section id="scaling-and-normalization" class="level4">
<h4 class="anchored" data-anchor-id="scaling-and-normalization">SCALING AND NORMALIZATION</h4>
<p><strong>Scaling and normalization are data preprocessing techniques used to transform the features or variables in a dataset to make them suitable for machine learning algorithms.</strong></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb117"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb117-1"><a href="#cb117-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb117-2"><a href="#cb117-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(palmerpenguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'palmerpenguins'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:modeldata':

    penguins</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(penguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The below code selects the numeric columns from the penguins dataset and normalizes it using scale() function. Finally, the as.data.frame() function converts the normalized dataset into a data frame and a species column from the original dataset is added to it. This normalized data frame can be used for ML algorithms and analysis that require standardized intputs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="co"># select only numeric columns</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>penguins_numeric <span class="ot">&lt;-</span> <span class="fu">select</span>(penguins, bill_length_mm, bill_depth_mm, flipper_length_mm, body_mass_g)</span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a><span class="co"># normalize numeric columns</span></span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>penguins_norm <span class="ot">&lt;-</span> <span class="fu">scale</span>(penguins_numeric)</span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a><span class="co"># convert back to data frame and add species column</span></span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a>peng.norm <span class="ot">&lt;-</span> <span class="fu">cbind</span>(<span class="fu">as.data.frame</span>(penguins_norm), <span class="at">species =</span> penguins<span class="sc">$</span>species)</span>
<span id="cb121-9"><a href="#cb121-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-10"><a href="#cb121-10" aria-hidden="true" tabindex="-1"></a><span class="co"># because scale() takes numeric matrix as input, we first remove Species column, then use cbind() to add the column back after normalization.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Summary of penguins dataset</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(penguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      species          island    bill_length_mm  bill_depth_mm  
 Adelie   :152   Biscoe   :168   Min.   :32.10   Min.   :13.10  
 Chinstrap: 68   Dream    :124   1st Qu.:39.23   1st Qu.:15.60  
 Gentoo   :124   Torgersen: 52   Median :44.45   Median :17.30  
                                 Mean   :43.92   Mean   :17.15  
                                 3rd Qu.:48.50   3rd Qu.:18.70  
                                 Max.   :59.60   Max.   :21.50  
                                 NA's   :2       NA's   :2      
 flipper_length_mm  body_mass_g       sex           year     
 Min.   :172.0     Min.   :2700   female:165   Min.   :2007  
 1st Qu.:190.0     1st Qu.:3550   male  :168   1st Qu.:2007  
 Median :197.0     Median :4050   NA's  : 11   Median :2008  
 Mean   :200.9     Mean   :4202                Mean   :2008  
 3rd Qu.:213.0     3rd Qu.:4750                3rd Qu.:2009  
 Max.   :231.0     Max.   :6300                Max.   :2009  
 NA's   :2         NA's   :2                                 </code></pre>
</div>
</div>
<p>Summary of the peng.norm data frame</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb124"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb124-1"><a href="#cb124-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(peng.norm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> bill_length_mm     bill_depth_mm      flipper_length_mm  body_mass_g     
 Min.   :-2.16535   Min.   :-2.05144   Min.   :-2.0563   Min.   :-1.8726  
 1st Qu.:-0.86031   1st Qu.:-0.78548   1st Qu.:-0.7762   1st Qu.:-0.8127  
 Median : 0.09672   Median : 0.07537   Median :-0.2784   Median :-0.1892  
 Mean   : 0.00000   Mean   : 0.00000   Mean   : 0.0000   Mean   : 0.0000  
 3rd Qu.: 0.83854   3rd Qu.: 0.78430   3rd Qu.: 0.8594   3rd Qu.: 0.6836  
 Max.   : 2.87166   Max.   : 2.20217   Max.   : 2.1395   Max.   : 2.6164  
 NA's   :2          NA's   :2          NA's   :2         NA's   :2        
      species   
 Adelie   :152  
 Chinstrap: 68  
 Gentoo   :124  
                
                
                
                </code></pre>
</div>
</div>
<p>The below code selects the max and min values from the penguins dataset excluding the species column (since it has characters in it) and also while avoiding NA values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a>max <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">select</span>(penguins, <span class="sc">-</span>species), <span class="dv">2</span>, max, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a>min <span class="ot">&lt;-</span> <span class="fu">apply</span>(<span class="fu">select</span>(penguins, <span class="sc">-</span>species), <span class="dv">2</span>, min, <span class="at">na.rm=</span><span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb127"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb127-1"><a href="#cb127-1" aria-hidden="true" tabindex="-1"></a>max</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           island    bill_length_mm     bill_depth_mm flipper_length_mm 
      "Torgersen"            "59.6"            "21.5"             "231" 
      body_mass_g               sex              year 
           "6300"            "male"            "2009" </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a>min</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           island    bill_length_mm     bill_depth_mm flipper_length_mm 
         "Biscoe"            "32.1"            "13.1"             "172" 
      body_mass_g               sex              year 
           "2700"          "female"            "2007" </code></pre>
</div>
</div>
<p>The below code utilizes the lapply() function to apply a function to each column of the “penguins_numeric” dataset. The function (x - min(x, na.rm = TRUE)) / (max(x, na.rm = TRUE) - min(x, na.rm = TRUE)) performs a min-max scaling on each column. The penguin_scaled dataset is binded with the species column using the cbind() function. FInally, the summary of the data frame is displayed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="co"># min-max normalization</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>penguin_scaled <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">lapply</span>(penguins_numeric, <span class="cf">function</span>(x) (x <span class="sc">-</span> <span class="fu">min</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)) <span class="sc">/</span> (<span class="fu">max</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>) <span class="sc">-</span> <span class="fu">min</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))))</span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb131-4"><a href="#cb131-4" aria-hidden="true" tabindex="-1"></a>penguin_scaled <span class="ot">&lt;-</span> <span class="fu">cbind</span>(penguins_norm, <span class="at">species =</span> penguins<span class="sc">$</span>species)</span>
<span id="cb131-5"><a href="#cb131-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb131-6"><a href="#cb131-6" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(penguin_scaled)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> bill_length_mm     bill_depth_mm      flipper_length_mm  body_mass_g     
 Min.   :-2.16535   Min.   :-2.05144   Min.   :-2.0563   Min.   :-1.8726  
 1st Qu.:-0.86031   1st Qu.:-0.78548   1st Qu.:-0.7762   1st Qu.:-0.8127  
 Median : 0.09672   Median : 0.07537   Median :-0.2784   Median :-0.1892  
 Mean   : 0.00000   Mean   : 0.00000   Mean   : 0.0000   Mean   : 0.0000  
 3rd Qu.: 0.83854   3rd Qu.: 0.78430   3rd Qu.: 0.8594   3rd Qu.: 0.6836  
 Max.   : 2.87166   Max.   : 2.20217   Max.   : 2.1395   Max.   : 2.6164  
 NA's   :2          NA's   :2          NA's   :2         NA's   :2        
    species     
 Min.   :1.000  
 1st Qu.:1.000  
 Median :2.000  
 Mean   :1.919  
 3rd Qu.:3.000  
 Max.   :3.000  
                </code></pre>
</div>
</div>
<p>Here, the Boston dataset from the package “MASS” is loaded and the summary of the age column in the dataset is displayed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(Boston, <span class="at">package=</span><span class="st">"MASS"</span>)</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Boston<span class="sc">$</span>age)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
   2.90   45.02   77.50   68.57   94.08  100.00 </code></pre>
</div>
</div>
<p>In the below code, the binning function is used to bin the values in the “age” column into 5 equal width bins and add them to the new “newAge” column. Finally, the summary of the newAge column is displayed. Binning is a pre-processing technique that groups a series of numerical data into sets of bins.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a>Boston<span class="sc">$</span>newAge <span class="ot">&lt;-</span> dlookr<span class="sc">::</span><span class="fu">binning</span>(Boston<span class="sc">$</span>age, <span class="dv">5</span>, <span class="at">type =</span> <span class="st">"equal"</span>) <span class="co">#create 5 bins and add new column newAge to Boston</span></span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Boston<span class="sc">$</span>newAge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         levels freq       rate
1   [2.9,22.32]   45 0.08893281
2 (22.32,41.74]   71 0.14031621
3 (41.74,61.16]   70 0.13833992
4 (61.16,80.58]   81 0.16007905
5   (80.58,100]  239 0.47233202</code></pre>
</div>
</div>
<p>This code is similar to the above code but labels are added to each bin created using labels argument in the binning() function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a>Boston<span class="sc">$</span>newAge <span class="ot">&lt;-</span> dlookr<span class="sc">::</span><span class="fu">binning</span>(Boston<span class="sc">$</span>age, <span class="at">nbins =</span> <span class="dv">5</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"very-young"</span>, <span class="st">"young"</span>, <span class="st">"mid"</span>, <span class="st">"older"</span>, <span class="st">"very-old"</span>), <span class="at">type =</span> <span class="st">"equal"</span>) <span class="co">#add labels</span></span>
<span id="cb137-2"><a href="#cb137-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb137-3"><a href="#cb137-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(Boston<span class="sc">$</span>newAge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      levels freq       rate
1 very-young   45 0.08893281
2      young   71 0.14031621
3        mid   70 0.13833992
4      older   81 0.16007905
5   very-old  239 0.47233202</code></pre>
</div>
</div>
<p>After running this code, we will have a modified “Boston” dataset with a new “newAge” column containing the binning results based on equal-depth bins. The table will show the distribution of observations across these bins, providing insights into how the “age” values are distributed in the dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"Hmisc"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into 'C:/Users/Joel Jojo/AppData/Local/R/win-library/4.3'
(as 'lib' is unspecified)</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in contrib.url(repos, "source"): trying to use CRAN without setting a mirror</code></pre>
</div>
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Hmisc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Hmisc'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:parsnip':

    translate</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:dplyr':

    src, summarize</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dlookr':

    describe</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    format.pval, units</code></pre>
</div>
<div class="sourceCode cell-code" id="cb148"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb148-1"><a href="#cb148-1" aria-hidden="true" tabindex="-1"></a>Boston<span class="sc">$</span>newAge <span class="ot">&lt;-</span> <span class="fu">cut2</span>(Boston<span class="sc">$</span>age, <span class="at">g =</span> <span class="dv">5</span>) <span class="co">#create 5 equal-depth bins and add new column newAge to Boston</span></span>
<span id="cb148-2"><a href="#cb148-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb148-3"><a href="#cb148-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(Boston<span class="sc">$</span>newAge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
[ 2.9, 38.1) [38.1, 66.1) [66.1, 86.1) [86.1, 95.7) [95.7,100.0] 
         102          101          101          101          101 </code></pre>
</div>
</div>
<p>This code cuts the age column in the dataset into 5 bins where the function calculates the quantiles to create bins of equal depth. The factor() function is used to convert the binning results into a factor variable.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb150"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb150-1"><a href="#cb150-1" aria-hidden="true" tabindex="-1"></a>Boston<span class="sc">$</span>newAge <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">cut2</span>(Boston<span class="sc">$</span>age, <span class="at">g =</span> <span class="dv">5</span>), <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"very-young"</span>, <span class="st">"young"</span>, <span class="st">"mid"</span>, <span class="st">"older"</span>, <span class="st">"very-old"</span>))</span>
<span id="cb150-2"><a href="#cb150-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb150-3"><a href="#cb150-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(Boston<span class="sc">$</span>newAge)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
very-young      young        mid      older   very-old 
       102        101        101        101        101 </code></pre>
</div>
</div>
<p>The below code genrates a histogram of the age column. The break statement specifies the intervals for the histogram(here 0-101 with intervals of 10).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb152"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb152-1"><a href="#cb152-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(Boston<span class="sc">$</span>age, <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="dv">0</span>, <span class="dv">101</span>,<span class="at">by =</span> <span class="dv">10</span>)) <span class="co">#seq() gives the function for breaks. The age ranges from 0 – 101.</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Exercise-2_files/figure-html/unnamed-chunk-71-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>This code will generate a histogram plot that visualizes the distribution of ages in the “Boston” dataset. The x-axis will represent age ranges, and the y-axis will represent the frequency of observations in each age range. The bin width of 10 means that each bin will cover a range of 10 years.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb153"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb153-1"><a href="#cb153-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb153-2"><a href="#cb153-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb153-3"><a href="#cb153-3" aria-hidden="true" tabindex="-1"></a>Boston <span class="sc">|&gt;</span></span>
<span id="cb153-4"><a href="#cb153-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> age)) <span class="sc">+</span></span>
<span id="cb153-5"><a href="#cb153-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="at">binwidth =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="Exercise-2_files/figure-html/unnamed-chunk-72-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Five numeric values are added to the vector “data”.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb154"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb154-1"><a href="#cb154-1" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">20</span>, <span class="dv">30</span>, <span class="dv">50</span>, <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>’max(abs(data)) calculates the maximum absolute value in the “data” vector. It takes the absolute value of all elements in the vector, finds the maximum among them, and returns that value.</p>
<p>nchar(max(abs(data))) counts the number of digits in the maximum absolute value in the “data” vector.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb155"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb155-1"><a href="#cb155-1" aria-hidden="true" tabindex="-1"></a>(<span class="at">nDigits =</span> <span class="fu">nchar</span>(<span class="fu">max</span>(<span class="fu">abs</span>(data)))) <span class="co">#nchar counts the number of characters</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 3</code></pre>
</div>
</div>
<p>The below code calculates a decimal scale for the “data” vector. It divides each element of the “data” vector by 10 raised to the power of “nDigits.”</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb157"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb157-1"><a href="#cb157-1" aria-hidden="true" tabindex="-1"></a>(<span class="at">decimalScale =</span> data <span class="sc">/</span> (<span class="dv">10</span><span class="sc">^</span>nDigits))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.01 0.02 0.03 0.05 0.10</code></pre>
</div>
</div>
<p>The matrix() function turns the age vector into a matrix. The number of rows is calculated by dividing the length of the age vector by 5, creating a matrix with 3 rows.</p>
<p>byrow= TRUE indicates that the values should be filled into the matrix by rows.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb159"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb159-1"><a href="#cb159-1" aria-hidden="true" tabindex="-1"></a>age <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">13</span>, <span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">16</span>, <span class="dv">19</span>, <span class="dv">20</span>, <span class="dv">20</span>, <span class="dv">21</span>, <span class="dv">22</span>, <span class="dv">22</span>, <span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">30</span>)</span>
<span id="cb159-2"><a href="#cb159-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb159-3"><a href="#cb159-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Separate data into bins of depth 3</span></span>
<span id="cb159-4"><a href="#cb159-4" aria-hidden="true" tabindex="-1"></a>(<span class="at">bins =</span> <span class="fu">matrix</span>(age, <span class="at">nrow =</span> <span class="fu">length</span>(age) <span class="sc">/</span> <span class="dv">5</span>, <span class="at">byrow =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3] [,4] [,5]
[1,]   13   15   16   16   19
[2,]   20   20   21   22   22
[3,]   25   25   25   25   30</code></pre>
</div>
</div>
<p>“bin_means” vector will contain the mean (average) value of each bin from the “bins” matrix. Each element of “bin_means” corresponds to the mean of a specific bin or group of values from the original “age” data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb161"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb161-1"><a href="#cb161-1" aria-hidden="true" tabindex="-1"></a>(<span class="at">bin_means =</span> <span class="fu">apply</span>(bins, <span class="dv">1</span>, <span class="at">FUN =</span> mean))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15.8 21.0 26.0</code></pre>
</div>
</div>
<p>The provided code sets each row in the “bins” matrix to have the mean value of its respective bin.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb163"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb163-1"><a href="#cb163-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(bins)) {</span>
<span id="cb163-2"><a href="#cb163-2" aria-hidden="true" tabindex="-1"></a>   bins[i,] <span class="ot">=</span> bin_means[i]</span>
<span id="cb163-3"><a href="#cb163-3" aria-hidden="true" tabindex="-1"></a> }</span>
<span id="cb163-4"><a href="#cb163-4" aria-hidden="true" tabindex="-1"></a>bins</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3] [,4] [,5]
[1,] 15.8 15.8 15.8 15.8 15.8
[2,] 21.0 21.0 21.0 21.0 21.0
[3,] 26.0 26.0 26.0 26.0 26.0</code></pre>
</div>
</div>
<p>This code calculates the mean of each bin in the “bins” matrix and then rounds those mean values to two decimal places. The result is stored in a vector named “age_bin_mean_smoothed.”</p>
<p>t(bins) transposes the “bins” matrix, swapping rows and columns, so that each column represents a bin.</p>
<p>as.vector() converts the transposed matrix into a vector, concatenating all the columns into a single vector.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb165"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb165-1"><a href="#cb165-1" aria-hidden="true" tabindex="-1"></a>(<span class="at">age_bin_mean_smoothed =</span> <span class="fu">round</span>(<span class="fu">as.vector</span>(<span class="fu">t</span>(bins)), <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 15.8 15.8 15.8 15.8 15.8 21.0 21.0 21.0 21.0 21.0 26.0 26.0 26.0 26.0 26.0</code></pre>
</div>
</div>
</section>
<section id="variable-correlations-and-dimensionality-reduction" class="level4">
<h4 class="anchored" data-anchor-id="variable-correlations-and-dimensionality-reduction"><strong>Variable correlations and dimensionality reduction</strong></h4>
<p><strong>Variable correlations help us understand relationships between variables, while dimensionality reduction techniques allow us to simplify complex datasets by reducing the number of variables while retaining meaningful information.</strong></p>
<p>‘test1 = chisq.test(racetable, correct=F)’ This line performs a chi-squared test of independence using the “racetable” matrix. ‘correct = F’ specifies that no correction for continuity should be applied.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb167"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb167-1"><a href="#cb167-1" aria-hidden="true" tabindex="-1"></a>racetable <span class="ot">=</span> <span class="fu">rbind</span>(<span class="fu">c</span>(<span class="dv">151</span>,<span class="dv">9</span>), <span class="fu">c</span>(<span class="dv">63</span>,<span class="dv">103</span>))</span>
<span id="cb167-2"><a href="#cb167-2" aria-hidden="true" tabindex="-1"></a>test1 <span class="ot">=</span> <span class="fu">chisq.test</span>(racetable, <span class="at">correct=</span>F)</span>
<span id="cb167-3"><a href="#cb167-3" aria-hidden="true" tabindex="-1"></a>test1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test

data:  racetable
X-squared = 115.01, df = 1, p-value &lt; 2.2e-16</code></pre>
</div>
</div>
<p>The provided code creates a four-dimensional array named “seniors” with specific data values, dimensions (‘dim = c(2, 2, 2, 2)’), and dimension names(dimnames=lost(“..” etc)).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb169"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb169-1"><a href="#cb169-1" aria-hidden="true" tabindex="-1"></a>seniors <span class="ot">&lt;-</span> <span class="fu">array</span>(<span class="at">data =</span> <span class="fu">c</span>(<span class="dv">911</span>, <span class="dv">44</span>, <span class="dv">538</span>, <span class="dv">456</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">43</span>, <span class="dv">279</span>, <span class="dv">911</span>, <span class="dv">44</span>, <span class="dv">538</span>, <span class="dv">456</span>, <span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">43</span>, <span class="dv">279</span>), </span>
<span id="cb169-2"><a href="#cb169-2" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dim =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>),</span>
<span id="cb169-3"><a href="#cb169-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="st">"cigarette"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"yes"</span>, <span class="st">"no"</span>),</span>
<span id="cb169-4"><a href="#cb169-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"marijuana"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"yes"</span>, <span class="st">"no"</span>),</span>
<span id="cb169-5"><a href="#cb169-5" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"alcohol"</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"yes"</span>, <span class="st">"no"</span>), </span>
<span id="cb169-6"><a href="#cb169-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="st">"age"</span> <span class="ot">=</span><span class="fu">c</span>(<span class="st">"younger"</span>, <span class="st">"older"</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb170"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb170-1"><a href="#cb170-1" aria-hidden="true" tabindex="-1"></a>seniors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>, , alcohol = yes, age = younger

         marijuana
cigarette yes  no
      yes 911 538
      no   44 456

, , alcohol = no, age = younger

         marijuana
cigarette yes  no
      yes   3  43
      no    2 279

, , alcohol = yes, age = older

         marijuana
cigarette yes  no
      yes 911 538
      no   44 456

, , alcohol = no, age = older

         marijuana
cigarette yes  no
      yes   3  43
      no    2 279</code></pre>
</div>
</div>
<p>seniors.tb &lt;- as.table(seniors): This line applies the as.table() function to the “seniors” array, converting it into a table format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb172"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb172-1"><a href="#cb172-1" aria-hidden="true" tabindex="-1"></a>seniors.tb <span class="ot">&lt;-</span> <span class="fu">as.table</span>(seniors)</span>
<span id="cb172-2"><a href="#cb172-2" aria-hidden="true" tabindex="-1"></a>seniors.tb</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>, , alcohol = yes, age = younger

         marijuana
cigarette yes  no
      yes 911 538
      no   44 456

, , alcohol = no, age = younger

         marijuana
cigarette yes  no
      yes   3  43
      no    2 279

, , alcohol = yes, age = older

         marijuana
cigarette yes  no
      yes 911 538
      no   44 456

, , alcohol = no, age = older

         marijuana
cigarette yes  no
      yes   3  43
      no    2 279</code></pre>
</div>
</div>
<p>The ‘seniors.df &lt;- as.data.frame(seniors.tb)’ line applies the as.data.frame() function to the “seniors.tb” table, converting it into a data frame format. Converting the table to a data frame format allows you to perform various data manipulation and analysis tasks.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb174"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb174-1"><a href="#cb174-1" aria-hidden="true" tabindex="-1"></a>seniors.df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(seniors.tb)</span>
<span id="cb174-2"><a href="#cb174-2" aria-hidden="true" tabindex="-1"></a>seniors.df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   cigarette marijuana alcohol     age Freq
1        yes       yes     yes younger  911
2         no       yes     yes younger   44
3        yes        no     yes younger  538
4         no        no     yes younger  456
5        yes       yes      no younger    3
6         no       yes      no younger    2
7        yes        no      no younger   43
8         no        no      no younger  279
9        yes       yes     yes   older  911
10        no       yes     yes   older   44
11       yes        no     yes   older  538
12        no        no     yes   older  456
13       yes       yes      no   older    3
14        no       yes      no   older    2
15       yes        no      no   older   43
16        no        no      no   older  279</code></pre>
</div>
</div>
<p>Freq ~ (cigarette * marijuana * alcohol * age) specifies the formula for the model.</p>
<p>It suggests that the response variable “Freq” is regressed on the interaction of three categorical predictor variables: “cigarette,” “marijuana,” “alcohol” and “age” .</p>
<p>data = seniors.df specifies the data frame from which the variables are taken.</p>
<p>family=poisson indicates that a Poisson distribution should be used for modeling the response variable.</p>
<p>The glm() function is used for fitting generalized linear models to a dataset.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb176"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb176-1"><a href="#cb176-1" aria-hidden="true" tabindex="-1"></a>mod.S4 <span class="ot">&lt;-</span> <span class="fu">glm</span>(Freq <span class="sc">~</span> (cigarette <span class="sc">*</span> marijuana <span class="sc">*</span> alcohol <span class="sc">*</span> age), <span class="at">data =</span> seniors.df, <span class="at">family=</span>poisson)</span>
<span id="cb176-2"><a href="#cb176-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mod.S4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = Freq ~ (cigarette * marijuana * alcohol * age), 
    family = poisson, data = seniors.df)

Coefficients:
                                             Estimate Std. Error z value
(Intercept)                                 6.815e+00  3.313e-02 205.682
cigaretteno                                -3.030e+00  1.544e-01 -19.633
marijuanano                                -5.267e-01  5.437e-02  -9.686
alcoholno                                  -5.716e+00  5.783e-01  -9.884
ageolder                                    3.219e-15  4.685e-02   0.000
cigaretteno:marijuanano                     2.865e+00  1.670e-01  17.159
cigaretteno:alcoholno                       2.625e+00  9.258e-01   2.835
marijuanano:alcoholno                       3.189e+00  5.996e-01   5.319
cigaretteno:ageolder                       -6.018e-15  2.183e-01   0.000
marijuanano:ageolder                       -3.402e-15  7.690e-02   0.000
alcoholno:ageolder                         -3.401e-15  8.178e-01   0.000
cigaretteno:marijuanano:alcoholno          -5.895e-01  9.424e-01  -0.626
cigaretteno:marijuanano:ageolder            6.541e-15  2.361e-01   0.000
cigaretteno:alcoholno:ageolder              1.682e-15  1.309e+00   0.000
marijuanano:alcoholno:ageolder              3.114e-15  8.480e-01   0.000
cigaretteno:marijuanano:alcoholno:ageolder -2.863e-15  1.333e+00   0.000
                                           Pr(&gt;|z|)    
(Intercept)                                 &lt; 2e-16 ***
cigaretteno                                 &lt; 2e-16 ***
marijuanano                                 &lt; 2e-16 ***
alcoholno                                   &lt; 2e-16 ***
ageolder                                    1.00000    
cigaretteno:marijuanano                     &lt; 2e-16 ***
cigaretteno:alcoholno                       0.00458 ** 
marijuanano:alcoholno                      1.04e-07 ***
cigaretteno:ageolder                        1.00000    
marijuanano:ageolder                        1.00000    
alcoholno:ageolder                          1.00000    
cigaretteno:marijuanano:alcoholno           0.53160    
cigaretteno:marijuanano:ageolder            1.00000    
cigaretteno:alcoholno:ageolder              1.00000    
marijuanano:alcoholno:ageolder              1.00000    
cigaretteno:marijuanano:alcoholno:ageolder  1.00000    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance:  5.7029e+03  on 15  degrees of freedom
Residual deviance: -2.2959e-13  on  0  degrees of freedom
AIC: 130.09

Number of Fisher Scoring iterations: 3</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb178"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb178-1"><a href="#cb178-1" aria-hidden="true" tabindex="-1"></a>mod.S3 <span class="ot">&lt;-</span> <span class="fu">glm</span>(Freq <span class="sc">~</span> (cigarette <span class="sc">*</span> marijuana <span class="sc">*</span> alcohol), <span class="at">data =</span> seniors.df, <span class="at">family =</span> poisson)</span>
<span id="cb178-2"><a href="#cb178-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mod.S3)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = Freq ~ (cigarette * marijuana * alcohol), family = poisson, 
    data = seniors.df)

Coefficients:
                                  Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)                        6.81454    0.02343 290.878  &lt; 2e-16 ***
cigaretteno                       -3.03035    0.10914 -27.765  &lt; 2e-16 ***
marijuanano                       -0.52668    0.03845 -13.699  &lt; 2e-16 ***
alcoholno                         -5.71593    0.40892 -13.978  &lt; 2e-16 ***
cigaretteno:marijuanano            2.86499    0.11806  24.267  &lt; 2e-16 ***
cigaretteno:alcoholno              2.62489    0.65466   4.010 6.08e-05 ***
marijuanano:alcoholno              3.18927    0.42400   7.522 5.40e-14 ***
cigaretteno:marijuanano:alcoholno -0.58951    0.66635  -0.885    0.376    
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 5.7029e+03  on 15  degrees of freedom
Residual deviance: 1.3367e-13  on  8  degrees of freedom
AIC: 114.09

Number of Fisher Scoring iterations: 3</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb180"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb180-1"><a href="#cb180-1" aria-hidden="true" tabindex="-1"></a>mod<span class="fl">.3</span> <span class="ot">&lt;-</span> <span class="fu">glm</span>(Freq <span class="sc">~</span> (cigarette <span class="sc">+</span> marijuana <span class="sc">+</span> alcohol)<span class="sc">^</span><span class="dv">2</span>, <span class="at">data =</span> seniors.df, <span class="at">family =</span> poisson)</span>
<span id="cb180-2"><a href="#cb180-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(mod<span class="fl">.3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = Freq ~ (cigarette + marijuana + alcohol)^2, family = poisson, 
    data = seniors.df)

Coefficients:
                        Estimate Std. Error z value Pr(&gt;|z|)    
(Intercept)              6.81387    0.02342 290.902   &lt;2e-16 ***
cigaretteno             -3.01575    0.10721 -28.130   &lt;2e-16 ***
marijuanano             -0.52486    0.03838 -13.674   &lt;2e-16 ***
alcoholno               -5.52827    0.31976 -17.289   &lt;2e-16 ***
cigaretteno:marijuanano  2.84789    0.11585  24.582   &lt;2e-16 ***
cigaretteno:alcoholno    2.05453    0.12308  16.692   &lt;2e-16 ***
marijuanano:alcoholno    2.98601    0.32858   9.088   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

(Dispersion parameter for poisson family taken to be 1)

    Null deviance: 5702.92195  on 15  degrees of freedom
Residual deviance:    0.74797  on  9  degrees of freedom
AIC: 112.83

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
<p>The cbind function combines these two sets of values into a single matrix or data frame where the original data columns and the fitted values are side by side. This can be useful for comparing the actual data to the model’s predictions and evaluating how well the model fits the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb182"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb182-1"><a href="#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cbind</span>(mod<span class="fl">.3</span><span class="sc">$</span>data, <span class="fu">fitted</span>(mod<span class="fl">.3</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   cigarette marijuana alcohol     age Freq fitted(mod.3)
1        yes       yes     yes younger  911     910.38317
2         no       yes     yes younger   44      44.61683
3        yes        no     yes younger  538     538.61683
4         no        no     yes younger  456     455.38317
5        yes       yes      no younger    3       3.61683
6         no       yes      no younger    2       1.38317
7        yes        no      no younger   43      42.38317
8         no        no      no younger  279     279.61683
9        yes       yes     yes   older  911     910.38317
10        no       yes     yes   older   44      44.61683
11       yes        no     yes   older  538     538.61683
12        no        no     yes   older  456     455.38317
13       yes       yes      no   older    3       3.61683
14        no       yes      no   older    2       1.38317
15       yes        no      no   older   43      42.38317
16        no        no      no   older  279     279.61683</code></pre>
</div>
</div>
<p>The code first removes rows with missing values and then calculates the correlation matrix for the remaining data. This can be useful for exploring the relationships between variables and identifying patterns in the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb184"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb184-1"><a href="#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr) <span class="co"># data manipulation</span></span>
<span id="cb184-2"><a href="#cb184-2" aria-hidden="true" tabindex="-1"></a>penguins_numeric <span class="sc">|&gt;</span> </span>
<span id="cb184-3"><a href="#cb184-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="sc">|&gt;</span></span>
<span id="cb184-4"><a href="#cb184-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">correlate</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 12 × 3
   var1              var2              coef_corr
   &lt;fct&gt;             &lt;fct&gt;                 &lt;dbl&gt;
 1 bill_depth_mm     bill_length_mm       -0.235
 2 flipper_length_mm bill_length_mm        0.656
 3 body_mass_g       bill_length_mm        0.595
 4 bill_length_mm    bill_depth_mm        -0.235
 5 flipper_length_mm bill_depth_mm        -0.584
 6 body_mass_g       bill_depth_mm        -0.472
 7 bill_length_mm    flipper_length_mm     0.656
 8 bill_depth_mm     flipper_length_mm    -0.584
 9 body_mass_g       flipper_length_mm     0.871
10 bill_length_mm    body_mass_g           0.595
11 bill_depth_mm     body_mass_g          -0.472
12 flipper_length_mm body_mass_g           0.871</code></pre>
</div>
</div>
<p>Initially, the rows are removed from the dataset, and columns species, island, sex are removed.</p>
<p>pca: The princomp() function is used to perform Principal Component Analysis on the “pca.data” data frame. This function calculates the principal components of the numeric variables in the data frame.</p>
<p>loadings(pca): This line extracts the loadings of the principal components from the PCA result. Loadings represent the coefficients of the original variables in the linear combinations that make up the principal components.</p>
<p>By examining the loadings, you can gain insights into which original variables have the most influence on each principal component.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb186"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb186-1"><a href="#cb186-1" aria-hidden="true" tabindex="-1"></a>pca.data <span class="ot">&lt;-</span> penguins <span class="sc">|&gt;</span></span>
<span id="cb186-2"><a href="#cb186-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>() <span class="sc">|&gt;</span></span>
<span id="cb186-3"><a href="#cb186-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>species, <span class="sc">-</span>island, <span class="sc">-</span>sex) </span>
<span id="cb186-4"><a href="#cb186-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb186-5"><a href="#cb186-5" aria-hidden="true" tabindex="-1"></a>pca <span class="ot">&lt;-</span> <span class="fu">princomp</span>(pca.data)</span>
<span id="cb186-6"><a href="#cb186-6" aria-hidden="true" tabindex="-1"></a><span class="fu">loadings</span>(pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Loadings:
                  Comp.1 Comp.2 Comp.3 Comp.4 Comp.5
bill_length_mm            0.319  0.941  0.108       
bill_depth_mm                    0.144 -0.984       
flipper_length_mm         0.943 -0.304 -0.125       
body_mass_g        1.000                            
year                                           0.997

               Comp.1 Comp.2 Comp.3 Comp.4 Comp.5
SS loadings       1.0    1.0    1.0    1.0    1.0
Proportion Var    0.2    0.2    0.2    0.2    0.2
Cumulative Var    0.2    0.4    0.6    0.8    1.0</code></pre>
</div>
</div>
<p>The head(pca$scores) command is used to display the first few rows of the scores matrix obtained from the Principal Component Analysis (PCA).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb188"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb188-1"><a href="#cb188-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(pca<span class="sc">$</span>scores) <span class="co"># pca result is a list, and the component scores are elements in the list</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Comp.1     Comp.2      Comp.3     Comp.4     Comp.5
[1,] -457.3251 -13.376298  1.24790414  0.3764738 -0.5892777
[2,] -407.2522  -9.205245 -0.03266674  1.0902167 -0.6935631
[3,] -957.0447   8.128321 -2.49146655 -0.7208225 -1.3723988
[4,] -757.1158   1.838910 -4.88056871 -2.0736684 -1.3477178
[5,] -557.1773  -3.416994 -1.12926657 -2.6292970 -1.1451328
[6,] -582.3093 -11.398030  0.80239704  1.1987261 -0.5976776</code></pre>
</div>
</div>
<p>pca$scores[,1:3]: This extracts the scores from the first three principal components ([,1:3]) obtained from the PCA analysis after the NA values are dropped.</p>
<p>Species = penguins_na$species: This adds the “Species” column from the original “penguins_na” data frame to the reduced data frame.</p>
<p>The resulting “peng.reduced” data frame retains information about the first three principal components and the “Species” column, which can be useful for further analysis or visualization of the reduced-dimensional data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb190"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb190-1"><a href="#cb190-1" aria-hidden="true" tabindex="-1"></a>penguins_na <span class="ot">&lt;-</span> penguins <span class="sc">|&gt;</span> </span>
<span id="cb190-2"><a href="#cb190-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">drop_na</span>()</span>
<span id="cb190-3"><a href="#cb190-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-4"><a href="#cb190-4" aria-hidden="true" tabindex="-1"></a>peng.reduced <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(pca<span class="sc">$</span>scores[,<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">Species =</span> penguins_na<span class="sc">$</span>species)</span>
<span id="cb190-5"><a href="#cb190-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb190-6"><a href="#cb190-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(peng.reduced)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Comp.1     Comp.2      Comp.3 Species
1 -457.3251 -13.376298  1.24790414  Adelie
2 -407.2522  -9.205245 -0.03266674  Adelie
3 -957.0447   8.128321 -2.49146655  Adelie
4 -757.1158   1.838910 -4.88056871  Adelie
5 -557.1773  -3.416994 -1.12926657  Adelie
6 -582.3093 -11.398030  0.80239704  Adelie</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb192"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb192-1"><a href="#cb192-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">"wavelets"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Installing package into 'C:/Users/Joel Jojo/AppData/Local/R/win-library/4.3'
(as 'lib' is unspecified)</code></pre>
</div>
<div class="cell-output cell-output-error">
<pre><code>Error in contrib.url(repos, "source"): trying to use CRAN without setting a mirror</code></pre>
</div>
<div class="sourceCode cell-code" id="cb195"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb195-1"><a href="#cb195-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(wavelets)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'wavelets'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:car':

    dwt</code></pre>
</div>
</div>
<p>wt &lt;- dwt(x, filter = “haar”, n.levels = 3): The dwt function performs the discrete wavelet transform (DWT) on the vector x.</p>
<p>filter = “haar”: Specifies the Haar wavelet as the wavelet filter to be used for the transformation.</p>
<p>n.levels = 3: Indicates that three levels of decomposition should be performed.</p>
<p>The result, stored in the wt variable, will include information about how the signal x is decomposed into different levels of approximation and detail coefficients.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb198"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb198-1"><a href="#cb198-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">4</span>, <span class="dv">4</span>)</span>
<span id="cb198-2"><a href="#cb198-2" aria-hidden="true" tabindex="-1"></a>wt <span class="ot">&lt;-</span> <span class="fu">dwt</span>(x,<span class="at">filter=</span><span class="st">"haar"</span>, <span class="at">n.levels =</span> <span class="dv">3</span>) <span class="co">#with 8-element vector, 3 level is the max.</span></span>
<span id="cb198-3"><a href="#cb198-3" aria-hidden="true" tabindex="-1"></a>wt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class "dwt"
Slot "W":
$W1
         [,1]
[1,] 0.000000
[2,] 1.414214
[3,] 1.414214
[4,] 0.000000

$W2
     [,1]
[1,]   -1
[2,]    0

$W3
         [,1]
[1,] 3.535534


Slot "V":
$V1
         [,1]
[1,] 2.828427
[2,] 1.414214
[3,] 5.656854
[4,] 5.656854

$V2
     [,1]
[1,]    3
[2,]    8

$V3
         [,1]
[1,] 7.778175


Slot "filter":
Filter Class: Daubechies
Name: HAAR
Length: 2
Level: 1
Wavelet Coefficients: 7.0711e-01 -7.0711e-01
Scaling Coefficients: 7.0711e-01 7.0711e-01


Slot "level":
[1] 3

Slot "n.boundary":
[1] 0 0 0

Slot "boundary":
[1] "periodic"

Slot "series":
     [,1]
[1,]    2
[2,]    2
[3,]    0
[4,]    2
[5,]    3
[6,]    5
[7,]    4
[8,]    4

Slot "class.X":
[1] "numeric"

Slot "attr.X":
list()

Slot "aligned":
[1] FALSE

Slot "coe":
[1] FALSE</code></pre>
</div>
</div>
<p>The idwt function is typically used to perform the inverse discrete wavelet transform (IDWT) to reconstruct the original signal from its wavelet coefficients.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb200"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb200-1"><a href="#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="fu">idwt</span>(wt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2 2 0 2 3 5 4 4</code></pre>
</div>
</div>
<p>The dwt function performs the discrete wavelet transform (DWT) on the vector x.</p>
<p>filter = wt.filter(c(0.5, -0.5)): Specifies a custom wavelet filter for the transformation. In this case, the filter is defined as [0.5, -0.5].</p>
<p>n.levels = 3: Indicates that three levels of decomposition should be performed.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb202"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb202-1"><a href="#cb202-1" aria-hidden="true" tabindex="-1"></a>xt <span class="ot">=</span> <span class="fu">dwt</span>(x, <span class="at">filter =</span> <span class="fu">wt.filter</span>(<span class="fu">c</span>(<span class="fl">0.5</span>, <span class="sc">-</span><span class="fl">0.5</span>)), <span class="at">n.levels =</span> <span class="dv">3</span>)</span>
<span id="cb202-2"><a href="#cb202-2" aria-hidden="true" tabindex="-1"></a>xt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class "dwt"
Slot "W":
$W1
     [,1]
[1,]    0
[2,]    1
[3,]    1
[4,]    0

$W2
     [,1]
[1,] -0.5
[2,]  0.0

$W3
     [,1]
[1,] 1.25


Slot "V":
$V1
     [,1]
[1,]    2
[2,]    1
[3,]    4
[4,]    4

$V2
     [,1]
[1,]  1.5
[2,]  4.0

$V3
     [,1]
[1,] 2.75


Slot "filter":
Filter Class: none
Name: NONE
Length: 2
Level: 1
Wavelet Coefficients: 5.0000e-01 -5.0000e-01
Scaling Coefficients: 5.0000e-01 5.0000e-01


Slot "level":
[1] 3

Slot "n.boundary":
[1] 0 0 0

Slot "boundary":
[1] "periodic"

Slot "series":
     [,1]
[1,]    2
[2,]    2
[3,]    0
[4,]    2
[5,]    3
[6,]    5
[7,]    4
[8,]    4

Slot "class.X":
[1] "numeric"

Slot "attr.X":
list()

Slot "aligned":
[1] FALSE

Slot "coe":
[1] FALSE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb204"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb204-1"><a href="#cb204-1" aria-hidden="true" tabindex="-1"></a><span class="fu">idwt</span>(xt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  0.3125  0.3125 -0.4375  0.5625  0.0000  1.0000  0.5000  0.5000</code></pre>
</div>
</div>
</section>
<section id="sampling" class="level4">
<h4 class="anchored" data-anchor-id="sampling">SAMPLING</h4>
<p><strong>Sampling is a statistical process of selecting a subset of individuals or items from a larger population or dataset for the purpose of making inferences or observations about the population.</strong></p>
<p>set.seed(1) sets the seed for the random number generator.</p>
<p>There is a vector called “age” with 12 age values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb206"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb206-1"><a href="#cb206-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>)</span>
<span id="cb206-2"><a href="#cb206-2" aria-hidden="true" tabindex="-1"></a>age <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">30</span>, <span class="dv">33</span>, <span class="dv">33</span>, <span class="dv">35</span>, <span class="dv">40</span>, <span class="dv">45</span>, <span class="dv">46</span>, <span class="dv">52</span>, <span class="dv">70</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The sample(age, 5) command selects a random sample of 5 values from the “age” vector. Since seed is set to using set.seed(1), the random sampling will be reproducible.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb207"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb207-1"><a href="#cb207-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample</span>(age, <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 45 30 35 25 25</code></pre>
</div>
</div>
<p>Sample() function selects 5 random values from the “age” vector with replacement. The replace = TRUE argument allows the same value to be selected more than once in the sample.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb209"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb209-1"><a href="#cb209-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sample</span>(age, <span class="dv">5</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 35 52 25 52 25</code></pre>
</div>
</div>
<p>By setting the seed and then summarizing the dataset, we ensure that the summary statistics are based on the same random operations each time you run the code.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb211"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb211-1"><a href="#cb211-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb211-2"><a href="#cb211-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">1</span>) <span class="co">#make results the same each run</span></span>
<span id="cb211-3"><a href="#cb211-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(algae)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    season       size       speed         mxPH            mnO2       
 autumn:40   large :44   high  :84   Min.   :5.600   Min.   : 1.500  
 spring:53   medium:84   low   :33   1st Qu.:7.705   1st Qu.: 7.825  
 summer:44   small :70   medium:81   Median :8.060   Median : 9.800  
 winter:61                           Mean   :8.019   Mean   : 9.135  
                                     3rd Qu.:8.400   3rd Qu.:10.800  
                                     Max.   :9.700   Max.   :13.400  
       Cl               NO3              NH4                oPO4       
 Min.   :  0.222   Min.   : 0.050   Min.   :    5.00   Min.   :  1.00  
 1st Qu.: 10.425   1st Qu.: 1.296   1st Qu.:   38.33   1st Qu.: 15.70  
 Median : 32.178   Median : 2.675   Median :  103.17   Median : 40.15  
 Mean   : 42.434   Mean   : 3.282   Mean   :  501.30   Mean   : 73.59  
 3rd Qu.: 57.492   3rd Qu.: 4.446   3rd Qu.:  226.95   3rd Qu.: 99.33  
 Max.   :391.500   Max.   :45.650   Max.   :24064.00   Max.   :564.60  
      PO4              Chla               a1               a2        
 Min.   :  1.00   Min.   :  0.200   Min.   : 0.000   Min.   : 0.000  
 1st Qu.: 41.38   1st Qu.:  2.000   1st Qu.: 1.525   1st Qu.: 0.000  
 Median :103.29   Median :  5.155   Median : 6.950   Median : 3.000  
 Mean   :137.88   Mean   : 13.355   Mean   :16.996   Mean   : 7.471  
 3rd Qu.:213.75   3rd Qu.: 17.200   3rd Qu.:24.800   3rd Qu.:11.275  
 Max.   :771.60   Max.   :110.456   Max.   :89.800   Max.   :72.600  
       a3               a4               a5               a6        
 Min.   : 0.000   Min.   : 0.000   Min.   : 0.000   Min.   : 0.000  
 1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.: 0.000  
 Median : 1.550   Median : 0.000   Median : 2.000   Median : 0.000  
 Mean   : 4.334   Mean   : 1.997   Mean   : 5.116   Mean   : 6.005  
 3rd Qu.: 4.975   3rd Qu.: 2.400   3rd Qu.: 7.500   3rd Qu.: 6.975  
 Max.   :42.800   Max.   :44.600   Max.   :44.400   Max.   :77.600  
       a7        
 Min.   : 0.000  
 1st Qu.: 0.000  
 Median : 1.000  
 Mean   : 2.487  
 3rd Qu.: 2.400  
 Max.   :31.600  </code></pre>
</div>
</div>
<p>This code first groups the “algae” dataset by the “season” variable using the group_by function. Then, it selects a random sample of 25% of the data within each season group using the sample_frac function.</p>
<p>This can be useful for exploring and analyzing a representative portion of the data within each season category.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb213"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb213-1"><a href="#cb213-1" aria-hidden="true" tabindex="-1"></a>sample <span class="ot">&lt;-</span>algae <span class="sc">|&gt;</span> <span class="fu">group_by</span>(season) <span class="sc">|&gt;</span> <span class="fu">sample_frac</span>(<span class="fl">0.25</span>)</span>
<span id="cb213-2"><a href="#cb213-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    season       size       speed         mxPH            mnO2       
 autumn:10   large :18   high  :16   Min.   :5.600   Min.   : 1.500  
 spring:13   medium:20   low   :13   1st Qu.:7.840   1st Qu.: 8.200  
 summer:11   small :11   medium:20   Median :8.100   Median : 9.700  
 winter:15                           Mean   :8.126   Mean   : 9.084  
                                     3rd Qu.:8.500   3rd Qu.:10.800  
                                     Max.   :9.700   Max.   :11.800  
       Cl               NO3             NH4               oPO4       
 Min.   :  0.222   Min.   :0.130   Min.   :   5.00   Min.   :  1.00  
 1st Qu.:  9.055   1st Qu.:0.921   1st Qu.:  43.75   1st Qu.: 21.08  
 Median : 28.333   Median :2.680   Median : 103.33   Median : 49.00  
 Mean   : 39.175   Mean   :2.794   Mean   : 213.60   Mean   : 78.49  
 3rd Qu.: 51.113   3rd Qu.:4.053   3rd Qu.: 225.00   3rd Qu.:125.67  
 Max.   :173.750   Max.   :6.640   Max.   :1495.00   Max.   :267.75  
      PO4              Chla               a1              a2       
 Min.   :  1.00   Min.   :  0.300   Min.   : 0.00   Min.   : 0.00  
 1st Qu.: 69.93   1st Qu.:  2.500   1st Qu.: 1.20   1st Qu.: 1.50  
 Median :125.80   Median :  6.209   Median : 5.70   Median : 5.60  
 Mean   :142.05   Mean   : 14.848   Mean   :17.94   Mean   :10.24  
 3rd Qu.:228.36   3rd Qu.: 20.829   3rd Qu.:18.10   3rd Qu.:14.50  
 Max.   :391.75   Max.   :110.456   Max.   :86.60   Max.   :72.60  
       a3               a4               a5               a6       
 Min.   : 0.000   Min.   : 0.000   Min.   : 0.000   Min.   : 0.00  
 1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.: 0.000   1st Qu.: 0.00  
 Median : 1.400   Median : 0.000   Median : 1.200   Median : 0.00  
 Mean   : 3.498   Mean   : 1.349   Mean   : 4.678   Mean   : 6.12  
 3rd Qu.: 4.000   3rd Qu.: 1.800   3rd Qu.: 7.700   3rd Qu.: 5.10  
 Max.   :24.800   Max.   :13.400   Max.   :22.100   Max.   :52.50  
       a7        
 Min.   : 0.000  
 1st Qu.: 0.000  
 Median : 1.000  
 Mean   : 2.367  
 3rd Qu.: 2.200  
 Max.   :31.200  </code></pre>
</div>
</div>
<p>The first line defines a vector called “age” containing a list of ages. Then the kmeans function performs k-means clustering on the “age” vector, where 3 is specified as the number of clusters to form.</p>
<p>s$cluster: This extracts the cluster assignments from the result of the k-means clustering and stores them in the “cluster” vector.</p>
<p>The result is a vector called “s$cluster” that contains cluster assignments for each age in the “age” vector.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb215"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb215-1"><a href="#cb215-1" aria-hidden="true" tabindex="-1"></a>age <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">13</span>, <span class="dv">15</span>, <span class="dv">16</span>, <span class="dv">16</span>, <span class="dv">19</span>, <span class="dv">20</span>, <span class="dv">20</span>, <span class="dv">21</span>, <span class="dv">22</span>, <span class="dv">22</span>, <span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">25</span>, <span class="dv">30</span>, <span class="dv">33</span>, <span class="dv">33</span>, <span class="dv">35</span>, <span class="dv">35</span>, <span class="dv">35</span>, <span class="dv">35</span>, <span class="dv">36</span>, <span class="dv">40</span>, <span class="dv">45</span>, <span class="dv">46</span>, <span class="dv">52</span>, <span class="dv">70</span>)</span>
<span id="cb215-2"><a href="#cb215-2" aria-hidden="true" tabindex="-1"></a>s <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(age, <span class="dv">3</span>) <span class="co">#cluster on age to form 3 clusters</span></span>
<span id="cb215-3"><a href="#cb215-3" aria-hidden="true" tabindex="-1"></a>s<span class="sc">$</span>cluster</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 3 3 3 3 3 2 2 2 2 2 2 2 2 2 2 1 1 1 1 1 1 1 1 1 1 1 1</code></pre>
</div>
</div>
<p>ageframe$condition &lt;- s$cluster: This adds a new column called “condition” to the “ageframe” data frame. The values in this column are assigned based on the cluster labels obtained from the k-means clustering performed earlier using the “s$cluster” vector.</p>
<p>cluster(ageframe, clustername = “condition”, size = 2): This line uses the cluster function to select 2 clusters out of the 3 based on the “condition” column.</p>
<p>The result is a subset of the data frame “ageframe,” containing data points belonging to two of the three clusters obtained from the k-means clustering.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb217"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb217-1"><a href="#cb217-1" aria-hidden="true" tabindex="-1"></a>ageframe <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(age)</span>
<span id="cb217-2"><a href="#cb217-2" aria-hidden="true" tabindex="-1"></a>ageframe<span class="sc">$</span>condition <span class="ot">&lt;-</span> s<span class="sc">$</span>cluster <span class="co"># add cluster label as condition</span></span>
<span id="cb217-3"><a href="#cb217-3" aria-hidden="true" tabindex="-1"></a><span class="fu">cluster</span>(ageframe, <span class="at">clustername =</span> <span class="st">"condition"</span>, <span class="at">size =</span> <span class="dv">2</span>) <span class="co"># select 2 clusters out of the three</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in cluster(ageframe, clustername = "condition", size = 2): the method
is not specified; by default, the method is srswor</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>   condition ID_unit      Prob
1          1      16 0.6666667
2          1      17 0.6666667
3          1      18 0.6666667
4          1      19 0.6666667
5          1      20 0.6666667
6          1      21 0.6666667
7          1      22 0.6666667
8          1      23 0.6666667
9          1      24 0.6666667
10         1      25 0.6666667
11         1      26 0.6666667
12         1      27 0.6666667
13         3       1 0.6666667
14         3       4 0.6666667
15         3       5 0.6666667
16         3       2 0.6666667
17         3       3 0.6666667</code></pre>
</div>
</div>
<p>Unable to recreate the <strong>Handling Text Datasets</strong> since I’m getting error “Error in DirSource(”Documents”): empty directory”</p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>